






<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="last-modified" content="2022-06-16 10:59:19 -0400">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <!-- meta "search-domain" used for google site search function google_search() -->
    <meta name="search-domain" value="">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/lesson.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/syntax.css" />
    
    <link rel="license" href="#license-info" />

    



    <!-- Favicons for everyone -->
    <link rel="apple-touch-icon-precomposed" sizes="57x57" href="/assets/favicons/swc/apple-touch-icon-57x57.png" />
    <link rel="apple-touch-icon-precomposed" sizes="114x114" href="/assets/favicons/swc/apple-touch-icon-114x114.png" />
    <link rel="apple-touch-icon-precomposed" sizes="72x72" href="/assets/favicons/swc/apple-touch-icon-72x72.png" />
    <link rel="apple-touch-icon-precomposed" sizes="144x144" href="/assets/favicons/swc/apple-touch-icon-144x144.png" />
    <link rel="apple-touch-icon-precomposed" sizes="60x60" href="/assets/favicons/swc/apple-touch-icon-60x60.png" />
    <link rel="apple-touch-icon-precomposed" sizes="120x120" href="/assets/favicons/swc/apple-touch-icon-120x120.png" />
    <link rel="apple-touch-icon-precomposed" sizes="76x76" href="/assets/favicons/swc/apple-touch-icon-76x76.png" />
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="/assets/favicons/swc/apple-touch-icon-152x152.png" />
    <link rel="icon" type="image/png" href="/assets/favicons/swc/favicon-196x196.png" sizes="196x196" />
    <link rel="icon" type="image/png" href="/assets/favicons/swc/favicon-96x96.png" sizes="96x96" />
    <link rel="icon" type="image/png" href="/assets/favicons/swc/favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="/assets/favicons/swc/favicon-16x16.png" sizes="16x16" />
    <link rel="icon" type="image/png" href="/assets/favicons/swc/favicon-128.png" sizes="128x128" />
    <meta name="application-name" content="Software Carpentry - DiscovR Workshop Curriculum"/>
    <meta name="msapplication-TileColor" content="#FFFFFF" />
    <meta name="msapplication-TileImage" content="/assets/favicons/swc/mstile-144x144.png" />
    <meta name="msapplication-square70x70logo" content="/assets/favicons/swc/mstile-70x70.png" />
    <meta name="msapplication-square150x150logo" content="/assets/favicons/swc/mstile-150x150.png" />
    <meta name="msapplication-wide310x150logo" content="/assets/favicons/swc/mstile-310x150.png" />
    <meta name="msapplication-square310x310logo" content="/assets/favicons/swc/mstile-310x310.png" />


    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
	<script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
	<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
	<![endif]-->

  <title>
  DiscovR Workshop Curriculum
  </title>  

  </head>
  <body>

    


<div class="panel panel-default life-cycle">
  <div id="life-cycle" class="panel-body pre-alpha">
    This lesson is still being designed and assembled (Pre-Alpha version)
  </div>
</div>





    <div class="container">
      
















  
  










<nav class="navbar navbar-default">
  <div class="container-fluid">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

      
      
      <a href="https://software-carpentry.org" class="pull-left">
        <img class="navbar-logo" src="/assets/img/swc-icon-blue.svg" alt="Software Carpentry logo" />
      </a>
      

      
      <a class="navbar-brand" href="/index.html">Home</a>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">

	
        <li><a href="/CODE_OF_CONDUCT.html">Code of Conduct</a></li>

        
	
        <li><a href="/setup.html">Setup</a></li>

        
        
        <li class="dropdown">
          <a href="/" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Episodes <span class="caret"></span></a>
          <ul class="dropdown-menu">
            
            
            <li><a href="/00-intro/index.html">Introduction to The DiscovR Workshop</a></li>
            
            
            <li><a href="/01-r-getting-started/index.html">Getting Started with R</a></li>
            
            
            <li><a href="/02-r-plotting/index.html">R for Plotting</a></li>
            
            
            <li><a href="/03-r-data-analysis/index.html">R for Data Analysis</a></li>
            
            
            <li><a href="/04-r-data-cleaning-merging/index.html">R for Data Cleaning and Merging</a></li>
            
            
            <li><a href="/05-r-markdown/index.html">Writing Reports with R Markdown</a></li>
            
            
            <li><a href="/06-conclusion/index.html">Conclusion</a></li>
            
	    <li role="separator" class="divider"></li>
            <li><a href="/aio/index.html">All in one page (Beta)</a></li>
          </ul>
        </li>
        
	

	
	
        <li class="dropdown">
          <a href="/" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Extras <span class="caret"></span></a>
          <ul class="dropdown-menu">
            <li><a href="/reference.html">Reference</a></li>
            
            
            
              <li><a href="/about/index.html">About</a></li>
            
            
            
            
              <li><a href="/guide/index.html">Instructor Notes</a></li>
            
            
          </ul>
        </li>
	

	
        <li><a href="/LICENSE.html">License</a></li>
	
	<li><a href="/edit//aio.md" data-checker-ignore>Improve this page <span class="glyphicon glyphicon-pencil" aria-hidden="true"></span></a></li>
	
      </ul>
      <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form>
    </div>
  </div>
</nav>







<h1 class="maintitle"><a href="/index.html">DiscovR Workshop Curriculum</a></h1>



<article>

<h1 id="introduction-to-the-discovr-workshop" class="maintitle">Introduction to The DiscovR Workshop</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 15 min
      <br />
      <strong>Exercises:</strong> 0 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>Who is the workshop for?</p>
</li>
	
	<li><p>What will the workshop cover?</p>
</li>
	
	<li><p>What else do I need to know about the workshop?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Set expectations.</p>
</li>
	
	<li><p>Meet each other.</p>
</li>
	
	<li><p>Introduce the workshop goals.</p>
</li>
	
	<li><p>Go over logistics.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<p><strong>DiscovR</strong> stands for “Data integration: strategies, concepts, and visuals in R”</p>

<h2 id="who-is-this-workshop-for">Who is this workshop for?</h2>

<p>This workshop is for data managers and others working with data who are interested in learning the foundations of data science and coding in R so you can use it in your own work. We believe everyone can learn to code, and that a lot of you will find it very useful for things such as data analysis and plotting.</p>

<p>This workshop is targeted to absolute beginners, and we expect that you have zero data science or coding experience coming in. 
That being said, you’re welcome to attend the workshop if you already have a coding background but want to learn more!</p>

<p>To provide an inclusive learning environment, we follow The Carpentries <a href="https://docs.carpentries.org/topic_folders/policies/code-of-conduct.html">Code of Conduct</a>. 
We expect that instructors, helpers, and learners abide by this code of conduct, including practicing the following behaviors:</p>

<ul>
  <li>Use welcoming and inclusive language.</li>
  <li>Be respectful of different viewpoints and experiences.</li>
  <li>Gracefully accept constructive criticism.</li>
  <li>Focus on what is best for the community.</li>
  <li>Show courtesy and respect towards other community members.</li>
</ul>

<h2 id="introducing-the-instructors-and-helpers">Introducing the instructors and helpers</h2>

<p>Now that you know a little about The Carpentries as an organization, the instructors and helpers will introduce themselves and what they’ll be teaching/helping with.</p>

<h2 id="introducing-participants">Introducing participants</h2>

<p>Introduce yourself with your preferred name, role, affiliation, work/research area, and Kenyan name and meaning.</p>

<h2 id="the-goal-of-the-workshop">The “goal” of the workshop</h2>

<p>Now that we all know each other, let’s learn a bit more about why we’re here. 
Our goal is to write a report to the United Nations on the relationship between GDP, life expectancy, and CO2 emissions. In other words, we are going to analyze  how countries’ economic strength or weakness may be related to  public health status and climate pollution, respectively.</p>

<p>To get to that point, we’ll need to learn how to manage data, make plots, and generate reports. The next section discusses in more detail exactly what we will cover.</p>

<h2 id="what-will-the-workshop-cover">What will the workshop cover?</h2>

<p>This workshop will introduce you to exploratory data analysis and effective data visualiation, and how to implement these concepts using the R programming language.</p>

<p>While we will focus primarily on public health applications, 
what you learn here are programs that are used everyday in computational 
<em>workflows</em> in diverse fields: 
microbiology, statistics, neuroscience, genetics, the social and behavioral 
sciences, such as psychology, economics, and many others.</p>

<p>A workflow is a set of steps to read data, analyze it, and produce numerical
and graphical results to support an assertion or hypothesis encapsulated into
a set of computer files that can be run from scratch on the same data to obtain
the same results. This is highly desirable in situations where the same work is 
done repeatedly – think of processing data from an annual
survey. It is
also desirable for <em>reproducibility</em>, which enables you and other people to look at
what you did and produce the same results later on. It is increasingly common for
people to publish scientific articles along with the data and  computer code that 
generated the results discussed within them.</p>

<p>The programs to be introduced are:</p>

<ol>
  <li><a href="../01-r-plotting"><em>R</em></a>: a statistical analysis and data management program,</li>
  <li><a href="../01-r-plotting"><em>RStudio</em></a>: a graphical interface to use R, and</li>
  <li><a href="../05-r-markdown"><em>R Markdown</em></a>: a method for writing reproducible reports. 
We’ll use these tools to manage data, perform basic statistical analyses, and make pretty plots!</li>
</ol>

<p>We will not try to make you an expert or even proficient with any of them, but we
hope to demonstrate the basics of controlling your code, automating your
work, and creating reproducible programs. We also hope to provide you with some
fundamentals that you can incorporate in your own work.</p>

<p>At the end, we provide links to resources you can use to learn about these
topics in more depth than this workshop can provide.</p>

<h2 id="asking-questions-and-getting-help">Asking questions and getting help</h2>

<p>One last note before we get into the workshop.</p>

<p>If you have general questions about a topic, please raise your hand to ask it. 
The instructor will definitely be willing to answer!</p>

<p>For more specific nitty-gritty questions about issues you’re having individually, we use <em>sticky notes</em> to indicate whether you are on track or need help. 
We’ll use these throughout the workshop to help us determine when you need help with a specific issue (a helper will come help), whether our pace is too fast, and whether you are finished with exercises. 
If you indicate that you need help because, for instance, you get an error in your code (e.g. red sticky), a helper will come help you figure things out.
Feel free to also call helpers over through a hand wave if we don’t see your sticky!</p>

<h2 id="other-miscellaneous-things">Other miscellaneous things</h2>

<p>If you’re in person, we’ll tell you where the bathrooms are!
Also let us know if there are any accommodations we can provide to help make your learning experience better.</p>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>We follow The Carpentries Code of Conduct.</p>
</li>
    
    <li><p>Our fundamental goal is to become more comfortable exploring and working with data.</p>
</li>
    
    <li><p>Our workshop goal is to write a sharable and reproducible report.</p>
</li>
    
    <li><p>This lesson content is targeted to absolute beginners with no data science or coding experience.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="getting-started-with-r" class="maintitle">Getting Started with R</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 90 min
      <br />
      <strong>Exercises:</strong> 15 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>What are R and R Studio?</p>
</li>
	
	<li><p>How do I perform tasks and store information?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>To become oriented with R and R Studio.</p>
</li>
	
	<li><p>To learn about functions and objects.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<h3 id="contents">Contents</h3>

<ol>
  <li><a href="#introduction-to-r-and-rstudio">Introduction to R and RStudio</a></li>
  <li><a href="#foundational-topics">Foundational topics</a>
    <ol>
      <li><a href="#functions">Functions</a></li>
      <li><a href="#objects">Objects</a></li>
    </ol>
  </li>
  <li><a href="#glossary-of-terms">Glossary of terms</a></li>
</ol>

<blockquote class="challenge">
  <h2 id="bonus-why-learn-to-program">Bonus: why learn to program?</h2>
  <p>Share why you’re interested in learning how to code.</p>
  <blockquote class="solution">
    <h2 id="solution">Solution:</h2>
    <p>There are lots of different reasons, including to perform data analysis and generate figures. I’m sure you have morespecific reasons for why you’d like to learn!</p>
  </blockquote>
</blockquote>

<h1 id="introduction-to-r-and-rstudio">Introduction to R and RStudio</h1>
<p><em><a href="#contents">Back to top</a></em></p>

<p>To perform exploratory analyses, we need the data we want to explore and a platform to analyze the data.</p>

<p>You already <a href="../setup.html">downloaded the data</a>. But what platform will we use to analyze the data? We have many options!</p>

<p>We could try to use a spreadsheet program like Microsoft Excel or Google sheets that have limited access, less flexibility, and don’t easily allow for things that are critical to <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285">“reproducible” research</a>, like easily sharing the steps used to explore and make changes to the original data.</p>

<p>We could also use a program like SAS or STATA, which are used by many epidemiologists. However, these programs are not freely available, the graphics are not as customizable, and there are not a ton of specialized packages for different niche analyses.</p>

<p>Instead, we’ll use a more general programming language to test our hypothesis. Today we will use R, but we could have also used Python for the same reasons we chose R. Both R and Python are freely available, the instructions you use to do the analysis are easily shared, and by using reproducible practices, it’s straightforward to add more data or to change settings like colors or the size of a plotting symbol.</p>

<blockquote class="solution">
  <h2 id="but-why-r-and-not-python">But why R and not Python?</h2>
  <p><a href="http://www.academichermit.com/2020/03/23/Why-R.html">There’s no great reason</a>.
Although there are subtle differences between the languages, it’s ultimately a matter of personal preference. Both are powerful and popular languages that have very well developed and welcoming communities of scientists that use them. As you learn more about R, you may find things that are annoying in R that aren’t so annoying in Python; the same could be said of learning Python. If the community you work in uses R, then you’re in the right place.</p>

</blockquote>

<p>To run R, all you really need is the R program, which is available for computers running the Windows, Mac OS X, or Linux operating systems. You downloaded R while getting <a href="../setup.html">set up</a> for this workshop.</p>

<p>To make your life in R easier, there is a great (and free!) program called RStudio that you also downloaded and used during <a href="../setup.html">set up</a>. As we work today, we’ll use features that are available in RStudio for writing and running code, managing projects, installing packages, getting help, and much more. It is important to remember that R and RStudio are different, but complementary programs. You need R to use RStudio.</p>

<p>To get started, we’ll spend a little time getting familiar with the RStudio environment and setting it up to suit your tastes. When you start RStudio, you’ll have three panels.</p>

<p><img src="../fig/r-plotting/initial_rstudio.png" width="700" /></p>

<p>On the left you’ll have a panel with three tabs - Console, Terminal, and Jobs. The Console tab is what running R from the command line looks like. This is where you can enter R code. Try typing in <code class="language-plaintext highlighter-rouge">2+2</code> at the prompt (&gt;). In the upper right panel are tabs indicating the Environment, History, and a few other things. If you click on the History tab, you’ll see the command you ran at the R prompt.</p>

<p><img src="../fig/r-plotting/history.png" width="700" /></p>

<p>In the lower right panel are tabs for Files, Plots, Packages, Help, and Viewer.</p>

<p>We’ll spend more time in each of these tabs as we go through the workshop, so we won’t spend a lot of time discussing them now.</p>

<p>Let’s get going on our analysis!</p>

<p>One of the helpful features in RStudio is the ability to create a project. A project is a special directory that contains all of the code and data that you will need to run an analysis.</p>

<p>At the top of your screen you’ll see the “File” menu. Select that menu and then the menu for “New Project…”.</p>

<p><img src="../fig/r-plotting/new_project_menu.png" width="600" /></p>

<p>When the smaller window opens, select “Existing Directory” and then the “Browse” button in the next window.</p>

<p><img src="../fig/r-plotting/existing_directory.png" width="600" /></p>

<p><img src="../fig/r-plotting/browse.png" width="600" /></p>

<p>Navigate to the directory that contains your code and data from the setup instructions and click the “Open” button.</p>

<p><img src="../fig/r-plotting/navigate_to_project.png" width="600" /></p>

<p>Then click the “Create Project” button.</p>

<p><img src="../fig/r-plotting/create_project.png" width="600" /></p>

<p>Did you notice anything change?</p>

<p>In the lower right corner of your RStudio session, you should notice that your
Files tab is now your project directory. You’ll also see a file called
un-report.Rproj in that directory.</p>

<p>From now on, you should start RStudio by double clicking on that file. This will
make sure you are in the correct directory when you run your analysis.</p>

<p><img src="../fig/r-plotting/files_with_rproj.png" width="600" /></p>

<p>We’d like to create a file where we can keep track of our R code.</p>

<p>Back in the “File” menu, you’ll see the first option is “New File”. Selecting “New File” opens another menu to the right and the first option is “R Script”. Select “R Script”.</p>

<p>Now we have a fourth panel in the upper left corner of RStudio that includes an <strong>Editor</strong> tab with an untitled R Script. Let’s save this file as <code class="language-plaintext highlighter-rouge">intro_to_r.R</code> in our project directory.</p>

<p>We will be entering R code into the <strong>Editor</strong> tab to run in our <strong>Console</strong> panel.</p>

<p>On line 1 of <code class="language-plaintext highlighter-rouge">intro_to_r.R</code>, type <code class="language-plaintext highlighter-rouge">2+2</code>.</p>

<p>With your cursor on the line with the <code class="language-plaintext highlighter-rouge">2+2</code>, press <kbd>Ctrl</kbd>+<kbd>Enter</kbd> on your keyboard. 
You should be able to see that <code class="language-plaintext highlighter-rouge">2+2</code> was run in the Console.
(You can also click <kbd>Run</kbd> in the top right side of the <strong>Editor</strong>, but this isn’t quite as easy.)</p>

<p>As you write more code, you can highlight multiple lines and then press <kbd>Ctrl</kbd>+<kbd>Enter</kbd>, orclick <kbd>Run</kbd>, to run all of the lines you have selected.</p>

<blockquote class="callout">
  <h2 id="comments">Comments</h2>
  <p>Sometimes you may want to write comments in your code to help you remember
what your code is doing, but you don’t want R to think these comments are a part
of the code you want to evaluate. That’s where <strong>comments</strong> come in! Anything
after a <code class="language-plaintext highlighter-rouge">#</code> symbol in your code will be ignored by R. For example, let’s say we
wanted to make a note of what each of the functions we just used do:</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># this is a comment</span><span class="w">
</span></code></pre></div>  </div>
</blockquote>

<h1 id="foundational-topics">Foundational topics</h1>

<h2 id="functions">Functions</h2>

<p>Functions are built-in procedures that automate a task for you.
You input arguments into a function and the function returns a value. 
We’ll go over a few math functions to get our feet wet.</p>

<p>You call a function in R by typing it’s name followed by opening then closing parenthesis. Each function has a purpose, which is often hinted at by the name of the function.</p>

<p>Let’s start with the <code class="language-plaintext highlighter-rouge">sqrt()</code> function.</p>

<p>Let’s try to run the function without anything inside the parenthesis.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nf">sqrt</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in sqrt(): 0 arguments passed to 'sqrt' which requires 1
</code></pre></div></div>

<p>We get an error message. Don’t panic! Error messages pop up all the time, and can be super helpful in debugging code.</p>

<p>In this case, the message tells us zero arguments were passed to the function, but we need to input at least one. 
Many functions, including <code class="language-plaintext highlighter-rouge">sqrt</code>, require additional pieces of information to do their job. 
We call these additional values “arguments” or “parameters.” 
You pass <strong>arguments</strong> to a function by placing values in between the parenthesis. 
A function takes in these arguments and works behind the scenes to output something we’re interested in.</p>

<p>For example, we want to provide a number to <code class="language-plaintext highlighter-rouge">sqrt()</code>, namely the number we want the square root of:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nf">sqrt</span><span class="p">(</span><span class="m">4</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 2
</code></pre></div></div>

<p>Here, the input argument is 4, and the output is 2, just like we’d expect.</p>

<p>Now let’s do an example where we might not know the expected output:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nf">sqrt</span><span class="p">(</span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 1.414214
</code></pre></div></div>

<p>Great, now let’s move onto a slightly more complicated function. 
If we want to round a number, we can use the <code class="language-plaintext highlighter-rouge">round()</code> function:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nf">round</span><span class="p">(</span><span class="m">3.14159</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 3
</code></pre></div></div>

<p>Why did this round to three? What if we want it to round to a different number of digits?</p>

<blockquote class="callout">
  <h2 id="pro-tip">Pro-tip</h2>

  <p>Each function has a help page that documents what arguments the function
expects and what value it will return. You can bring up the help page a few
different ways. If you have typed the function name in the <strong>Editor</strong> windows,
you can put your cursor on the function name and press <kbd>F1</kbd> to open
help page in the <strong>Help</strong> viewer in the lower right corner of RStudio. You can
also type <code class="language-plaintext highlighter-rouge">?</code> followed by the function name in the console.</p>

  <p>For example, try running <code class="language-plaintext highlighter-rouge">?round</code> in the console. A help page should pop up with
information about what the function is used for and how to use it, as well as
useful examples of the function in action. As you can see, 
<code class="language-plaintext highlighter-rouge">round()</code> has two arguments: the numeric input and the number of digits to round to.</p>

</blockquote>

<p>We can use the digits argument in <code class="language-plaintext highlighter-rouge">round()</code> to change how many decimal places are kept:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nf">round</span><span class="p">(</span><span class="m">3.14159</span><span class="p">,</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 3.14
</code></pre></div></div>

<p>Sometimes it is helpful - or even necessary - to include the argument name, but often we can skip the argument name, if the argument values are passed in the order they are defined:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nf">round</span><span class="p">(</span><span class="m">3.14159</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 3.14
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="position-of-the-arguments-in-functions">Position of the arguments in functions</h2>
  <p>Which of the following lines of code will give you an output of 3.14? For the one(s) that don’t give you 3.14, what do they give you?</p>

  <ol>
    <li><code class="language-plaintext highlighter-rouge">round(x = 3.1415)</code></li>
    <li><code class="language-plaintext highlighter-rouge">round(x = 3.1415, digits = 2)</code></li>
    <li><code class="language-plaintext highlighter-rouge">round(digits = 2, x = 3.1415)</code></li>
    <li><code class="language-plaintext highlighter-rouge">round(2, 3.1415)</code></li>
    <li><code class="language-plaintext highlighter-rouge">round(3.14159265, 2)</code></li>
  </ol>

  <blockquote class="solution">
    <h2 id="solution-1">Solution</h2>
    <ul>
      <li>The 1st line will give you 3 because the default number of digits is 0.</li>
      <li>The 2nd and 3rd lines will give you the right answer because the arguments are named, and when you use names the order doesn’t matter.</li>
      <li>The 4th line will give you 2 because, since you didn’t name the arguments, x=2 and digits=3.1415.</li>
      <li>The 5th line will also give you the right answer because the arguments are in the correct order.</li>
    </ul>
  </blockquote>
</blockquote>

<blockquote class="challenge">
  <h2 id="bonus-exercise-taking-logarithms">Bonus Exercise: taking logarithms</h2>
  <p>Calculate the following:</p>
  <ol>
    <li>Natural log (ln) of 10</li>
    <li>Log base 10 of 10 (challenge: try to do this 2 different ways), and</li>
    <li>Log base 3 of 10</li>
  </ol>

  <blockquote class="source solution">
    <h2 id="solution-2">Solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># natural log (ln) of 10</span><span class="w">
</span><span class="nf">log</span><span class="p">(</span><span class="m">10</span><span class="p">)</span><span class="w">
</span><span class="c1"># log base 10 of 10</span><span class="w">
</span><span class="n">log10</span><span class="p">(</span><span class="m">10</span><span class="p">)</span><span class="w">
</span><span class="nf">log</span><span class="p">(</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">base</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">)</span><span class="w">
</span><span class="c1"># log base 3 of 10</span><span class="w">
</span><span class="o">&gt;</span><span class="w"> </span><span class="nf">log</span><span class="p">(</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">base</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>
  </blockquote>
</blockquote>

<p>If all this function stuff sounds confusing, don’t worry! We’ll see a bunch of examples as we go that will make things clearer.</p>

<h2 id="objects">Objects</h2>

<p>Sometimes we want to store information for later use or manipulation. 
To do this in R, we store the information, or <em>object</em> , in a variable name that you can think of like a storage box.</p>

<p>Let’s say we want to round the square root of a number.
One way we can do this is to put a function inside a function:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nf">round</span><span class="p">(</span><span class="nf">sqrt</span><span class="p">(</span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 1.41
</code></pre></div></div>

<p>Another way is to store the square root output first, and then round that.</p>

<p>To store an object for later, we have to decide on a name of the box we want to store it in first. Let’s say we want to call it <code class="language-plaintext highlighter-rouge">square_root</code>.
Then we have to tell R what we want to put in the object name. We use the <code class="language-plaintext highlighter-rouge">&lt;-</code> symbol, which is the <strong>assignment operator</strong> to assign values generated or typed on the right to object names on the left. An alternative symbol that you might see used as an <strong>assignment operator</strong> is the <code class="language-plaintext highlighter-rouge">=</code> but it is clearer to only use <code class="language-plaintext highlighter-rouge">&lt;-</code> for assignment. We use this symbol so often that RStudio has a keyboard short cut for it: <kbd>Alt</kbd>+<kbd>-</kbd> on Windows, and <kbd>Option</kbd>+<kbd>-</kbd> on Mac.</p>

<p>Let’s assign <code class="language-plaintext highlighter-rouge">sqrt(2)</code> to the object <code class="language-plaintext highlighter-rouge">square_root</code>. We can see that <code class="language-plaintext highlighter-rouge">square_root</code> contains the square root of 2:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">square_root</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span><span class="m">2</span><span class="p">)</span><span class="w">
</span><span class="n">square_root</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 1.414214
</code></pre></div></div>

<p>In R terms, <code class="language-plaintext highlighter-rouge">square_root</code> is a named <strong>object</strong> that references or stores something. In this case, <code class="language-plaintext highlighter-rouge">square_root</code> stores the square root of 2.</p>

<p>Notice that we also have a new value in our environment in the upper right hand corner of RStudio.
This panel lists all of the objects that we have stored in our environment, 
it’s kind of like a view into our storage room (environment) of all the boxes (objects) of things we have access to.</p>

<p>Now let’s round the square root of 2 to 2 decimal places:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sqrt_rounded</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">square_root</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span><span class="n">sqrt_rounded</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 1.41
</code></pre></div></div>

<p>This is a fairly straightforward example, but you’ll see the usefulness of storing things in variables as the workshop progresses.</p>

<p>Now, what happens to <code class="language-plaintext highlighter-rouge">sqrt_rounded</code> if we update <code class="language-plaintext highlighter-rouge">square_root</code>?</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">square_root</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">sqrt</span><span class="p">(</span><span class="m">4</span><span class="p">)</span><span class="w">
</span><span class="n">square_root</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 2
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">sqrt_rounded</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] 1.41
</code></pre></div></div>

<p>It doesn’t update! That’s because we haven’t re-run the code that rounded <code class="language-plaintext highlighter-rouge">square_root</code>. 
The values don’t update automatically like in a spreadsheet.</p>

<blockquote class="challenge">
  <h2 id="predicting-object-contents">Predicting object contents</h2>
  <p>What is <code class="language-plaintext highlighter-rouge">my_number</code> after these three lines are run?</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">my_number</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">10</span><span class="w">
</span><span class="n">my_number</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">5</span><span class="w">
</span><span class="n">my_number</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">my_number</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="m">7</span><span class="w">
</span></code></pre></div>  </div>
  <ol>
    <li>10</li>
    <li>15</li>
    <li>17</li>
    <li>22</li>
  </ol>

  <blockquote class="source solution">
    <h2 id="solution-3">Solution</h2>
    <p>The answer is 17 because 10 is stored in <code class="language-plaintext highlighter-rouge">my_number</code> in the first line, 15 is printed after the second line but is not stored 
so <code class="language-plaintext highlighter-rouge">my_number</code> remains 10, and then 7 is added to <code class="language-plaintext highlighter-rouge">my_number</code> in the third line, making 17. If we ran the third line again,
<code class="language-plaintext highlighter-rouge">my_number</code> would be 24. Because the object value changes depending on the number of times we run the final line, 
in most cases it’s best practice to not overwrite objects like this.</p>
  </blockquote>
</blockquote>

<blockquote class="checklist">
  <h2 id="guidelines-on-naming-objects">Guidelines on naming objects</h2>
  <ul>
    <li>You want your object names to be explicit and not too long.</li>
    <li>They cannot start with a number (2x is not valid, but x2 is).</li>
    <li>R is case sensitive, so for example, weight is different from Weight.</li>
    <li>You cannot use spaces in the name.</li>
    <li>There are some names that cannot be used because they are the names of fundamental functions in R (e.g., if, else, for; see <a href="https://stat.ethz.ch/R-manual/R-devel/library/base/html/Reserved.html">here</a> for a complete list). If in doubt, check the help to see if the name is already in use (<code class="language-plaintext highlighter-rouge">?function_name</code>).</li>
    <li>It’s best to avoid dots (.) within names. Many function names in R itself have them and dots also have a special meaning (methods) in R and other programming languages.</li>
    <li>It is recommended to use nouns for object names and verbs for function names.</li>
    <li>Be consistent in the styling of your code, such as where you put spaces, how you name objects, etc. Using a consistent coding style makes your code clearer to read for your future self and your collaborators. One popular style guide can be found through the <a href="https://style.tidyverse.org/">tidyverse</a>.</li>
  </ul>
</blockquote>

<blockquote class="challenge">
  <h2 id="bonus-exercise-bad-names-for-objects">Bonus Exercise: Bad names for objects</h2>
  <p>Try to assign values to some new variable names. What do you notice? After running all four lines of code below, what value do you think the variable <code class="language-plaintext highlighter-rouge">Flower</code> holds?</p>

  <div class="language-r source highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="m">1</span><span class="n">number</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">3</span><span class="w">
</span><span class="n">Flower</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"marigold"</span><span class="w">
</span><span class="n">flower</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"rose"</span><span class="w">
</span><span class="n">favorite</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">12</span><span class="w">
</span></code></pre></div>  </div>

  <blockquote class="solution">
    <h2 id="solution-4">Solution</h2>
    <p>Notice that we get an error when we try to assign values to <code class="language-plaintext highlighter-rouge">1number</code> and <code class="language-plaintext highlighter-rouge">favorite number</code>. This is because we cannot start an object name with a numeral and we cannot have spaces in object names. The object <code class="language-plaintext highlighter-rouge">Flower</code> still holds “marigold.” This is because R is case-sensitive, so running <code class="language-plaintext highlighter-rouge">flower &lt;- "rose"</code> does NOT change the <code class="language-plaintext highlighter-rouge">Flower</code> object. This can get confusing, and is why we generally avoid having objects with the same name and different capitalization.</p>
  </blockquote>
</blockquote>

<blockquote class="callout">
  <h2 id="getting-unstuck">Getting unstuck</h2>
  <p>Sometimes you may accidentally run a line of code that isn’t quite complete yet. For instance:</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">my_number</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> 
</span></code></pre></div>  </div>
  <p>What happens when you run this? In your console at the bottom of your screen, you may see a <code class="language-plaintext highlighter-rouge">+</code>
instead of a <code class="language-plaintext highlighter-rouge">&gt;</code> at the beginning of the line. This means that R is waiting for more information.
In this case, it’s because it doesn’t know what you want to store in <code class="language-plaintext highlighter-rouge">my_number</code>. You can do one of
two things if this happens - finish the command you want to type (e.g. by entering a number), or
hit the escape key to get unstuck.</p>
</blockquote>

<blockquote class="callout">
  <h2 id="quotes-vs-no-quotes">Quotes vs. No Quotes</h2>
  <p>Let’s say we wanted to print out a word:</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">tree</span><span class="w">
</span></code></pre></div>  </div>

  <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in eval(expr, envir, enclos): object 'tree' not found
</code></pre></div>  </div>
  <p>You’ll notice that we get an error, that the object ‘tree’ is not found.
This is because R is looking for an object called <code class="language-plaintext highlighter-rouge">tree</code>. But what we really want is to just  print out the word “tree”. To do this, we put the word in quotes (single or double) so R knows that it’s not an object it needs to look for:</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="s2">"tree"</span><span class="w">
</span></code></pre></div>  </div>

  <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>[1] "tree"
</code></pre></div>  </div>
</blockquote>

<h1 id="glossary-of-terms">Glossary of terms</h1>
<p><em><a href="#contents">Back to top</a></em></p>

<ul>
  <li>Comments: lines or parts of lines that are not run. In R, comments start with a <code class="language-plaintext highlighter-rouge">#</code>.</li>
  <li>Function: takes input and generates output.</li>
  <li>Object: way to store information for later use and manipulation.</li>
</ul>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>R is a free programming language used by many for reproducible data analysis.</p>
</li>
    
    <li><p>Functions allow you to perform complex tasks.</p>
</li>
    
    <li><p>Objects allow you to store information.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="r-for-plotting" class="maintitle">R for Plotting</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 90 min
      <br />
      <strong>Exercises:</strong> 15 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How do I read data into R?</p>
</li>
	
	<li><p>What are geometries and aesthetics?</p>
</li>
	
	<li><p>How can I use R to create and save professional data visualizations?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>To be able to read in data from csv files.</p>
</li>
	
	<li><p>To create plots with both discrete and continuous variables.</p>
</li>
	
	<li><p>To understand mapping and layering using <code class="language-plaintext highlighter-rouge">ggplot2</code>.</p>
</li>
	
	<li><p>To be able to modify a plot’s color, theme, and axis labels.</p>
</li>
	
	<li><p>To be able to save plots to a local directory.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<h3 id="contents">Contents</h3>

<ol>
  <li><a href="#loading-and-reviewing-data">Loading and reviewing data</a></li>
  <li><a href="#our-first-plot">Our first plot</a>
    <ul>
      <li><a href="#creating-our-first-plot">Creating our first plot</a></li>
      <li><a href="#naming-our-first-plot">Saving our first plot</a></li>
    </ul>
  </li>
  <li><a href="#plotting-for-data-exploration">Plotting for data exploration</a>
    <ul>
      <li><a href="#importing-datasets">Importing datasets</a></li>
      <li><a href="#discrete-plots">Discrete plots</a></li>
      <li><a href="#adding-to-plot-objects">Adding to plot objects</a></li>
      <li><a href="#color-vs-fill">Color vs. Fill</a></li>
      <li><a href="#layers">Layers</a></li>
      <li><a href="#univariate-plots">Univariate plots</a>
        <ul>
          <li><a href="#univariate-continuous">Univariate continuous</a></li>
          <li><a href="#univariate-discrete">Univariate discrete</a></li>
        </ul>
      </li>
      <li><a href="#plot-themes">Plot themes</a></li>
      <li><a href="#facets">Facets</a></li>
    </ul>
  </li>
  <li><a href="#creating-and-saving-your-own-plot">Creating and saving your own plot</a></li>
  <li><a href="#applying-it-to-your-own-data">Applying it to your own data</a></li>
  <li><a href="#glossary-of-terms">Glossary of terms</a></li>
</ol>

<h1 id="loading-and-reviewing-data">Loading and reviewing data</h1>

<blockquote class="callout">
  <h2 id="the-tidyverse-vs-base-r">The tidyverse vs Base R</h2>
  <p>If you’ve used R before, you may have learned commands that are different than the ones we will be using during this workshop. We will be focusing on functions from the <a href="https://www.tidyverse.org/">tidyverse</a>. The “tidyverse” is a collection of R packages that have been designed to work well together and offer many convenient features that do not come with a fresh install of R (aka “base R”). These packages are very popular and have a lot of developer support including many staff members from RStudio. These functions generally help you to write code that is easier to read and maintain. We believe learning these tools will help you become more productive more quickly.</p>
</blockquote>

<p>First, we’re going to load the tidyverse package. Packages are useful because they contain pre-made functions to do specific tasks. Tidyverse contains a set of functions that makes it easier for us to do complex analyses and create professional visualizations in R. The way we access all of these useful functions functions is by running the following command:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<blockquote class="callout">
  <h2 id="whats-with-all-those-messages">What’s with all those messages???</h2>

  <p>When you loaded the <code class="language-plaintext highlighter-rouge">tidyverse</code> package, you probably got a message like the
one we got above. Don’t panic! These messages are just giving you more
information about what happened when you loaded <code class="language-plaintext highlighter-rouge">tidyverse</code>. The <code class="language-plaintext highlighter-rouge">tidyverse</code> is
actually a collection of several different packages, so the first section of the
message tells us what packages were installed when we loaded <code class="language-plaintext highlighter-rouge">tidyverse</code> (these
include <code class="language-plaintext highlighter-rouge">ggplot2</code>, which we’ll be using a lot in this lesson, and <code class="language-plaintext highlighter-rouge">dyplr</code>, which
you’ll be introduced to tomorrow in the 
<a href="../05-r-markdown">R for Data Analysis lesson</a>).</p>

  <p>The second section of messages gives a list of “conflicts.” Sometimes, the
same function name will be used in two different packages, and R has to decide
which function to use. For example, our message says that:</p>

  <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>dplyr::filter() masks stats::filter()
</code></pre></div>  </div>

  <p>This means that two different packages (<code class="language-plaintext highlighter-rouge">dyplr</code> from <code class="language-plaintext highlighter-rouge">tidyverse</code> and <code class="language-plaintext highlighter-rouge">stats</code>
from base R) have a function named <code class="language-plaintext highlighter-rouge">filter()</code>. By default, R uses the function
that was most recently loaded, so if we try using the <code class="language-plaintext highlighter-rouge">filter()</code> function after
loading <code class="language-plaintext highlighter-rouge">tidyverse</code>, we will be using the <code class="language-plaintext highlighter-rouge">filter()</code> function from <code class="language-plaintext highlighter-rouge">dplyr()</code>.</p>

</blockquote>

<p>Okay, now let’s read in our data, gapminder_1997.csv, which is in the data folder in our project directory. 
We’re going to use the <code class="language-plaintext highlighter-rouge">read_csv()</code> function that we loaded in with the tidyverse, 
and save it to <code class="language-plaintext highlighter-rouge">gapminder_1997</code>, which will act as a placeholder for our data.
This function takes a file path and returns a tibble, which is basically a table (that we sometimes call a data frame…).</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_1997</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/gapminder_1997.csv"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Rows: 142 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): country, continent
dbl (3): pop, lifeExp, gdpPercap

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
</code></pre></div></div>

<p>First, a few things printed out to the screen: it tells us how many rows and columns are in our data, 
and information about each of the columns. 
Each row contains life expectancy (“lifeExp”), the total population (“pop”), and the per capita gross domestic product (“gdpPercap”) for a given country (“country”).
There is also a column that says which continent each country is in (“continent”). Note that both North America and South America are combined into one category called “Americas”.
We can see that two of the columns are characters (categorical variables), and three are doubles (numbers).</p>

<p><strong>Note</strong> In anything before R 4.0, categorical variables used to be read in as factors, which are <a href="https://www.tutorialspoint.com/r/r_factors.htm">special data objects</a> that are used to store categorical data and have limited numbers of unique values. The unique values of a factor are tracked via the “levels” of a factor. A factor will always remember all of its levels even if the values don’t actually appear in your data. The factor will also remember the order of the levels and will always print values out in the same order (by default this order is alphabetical).</p>

<p>If your columns are stored as character values but you need factors for plotting, ggplot will convert them to factors for you as needed.</p>

<p>Now let’s look at the data a bit more. 
In the <strong>Environment</strong> tab in the upper right corner of RStudio, lou will now see <code class="language-plaintext highlighter-rouge">gapminder_1997</code> listed.
If you click on it, it will pop up in a tab next to your script.</p>

<p>After we’ve reviewed the data, you’ll want to make sure to click the tab in the upper left to return to your <code class="language-plaintext highlighter-rouge">gdp_population.R</code> file so we can start writing some code.</p>

<p>Another way to look at the data is to print it out to the console:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_1997</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 142 × 5
   country           pop continent lifeExp gdpPercap
   &lt;chr&gt;           &lt;dbl&gt; &lt;chr&gt;       &lt;dbl&gt;     &lt;dbl&gt;
 1 Afghanistan  22227415 Asia         41.8      635.
 2 Albania       3428038 Europe       73.0     3193.
 3 Algeria      29072015 Africa       69.2     4797.
 4 Angola        9875024 Africa       41.0     2277.
 5 Argentina    36203463 Americas     73.3    10967.
 6 Australia    18565243 Oceania      78.8    26998.
 7 Austria       8069876 Europe       77.5    29096.
 8 Bahrain        598561 Asia         73.9    20292.
 9 Bangladesh  123315288 Asia         59.4      973.
10 Belgium      10199787 Europe       77.5    27561.
# … with 132 more rows
</code></pre></div></div>

<p>The <code class="language-plaintext highlighter-rouge">read_csv()</code> function took the file path we provided, did who-knows-what behind the scenes, and then outputted an R object with the data stored in that csv file. All that, with one short line of code!</p>

<blockquote class="callout">
  <h2 id="data-objects">Data objects</h2>
  <p>There are many different ways to store data in R. Most objects have a table-like structure with rows and columns. We will refer to these objects generally as “data objects”. If you’ve used R before, you many be used to calling them “data.frames”. Functions from the “tidyverse” such as <code class="language-plaintext highlighter-rouge">read_csv()</code> work with objects called “tibbles”, which are a specialized kind of “data.frame.” Another common way to store data is a “data.table”. All of these types of data objects (tibbles, data.frames, and data.tables) can be used with the commands we will learn in this lesson to make plots. We may sometimes use these terms interchangeably.</p>
</blockquote>

<blockquote class="challenge">
  <h2 id="bonus-reading-in-an-excel-file">Bonus: Reading in an excel file</h2>
  <p>Say you have an excel file and not a csv - how would you read that in? Hint: Use the Internet to help you figure it out!</p>

  <blockquote class="source solution">
    <h2 id="solution">Solution</h2>
    <p>One way is using the <code class="language-plaintext highlighter-rouge">read_excel</code> function in the <code class="language-plaintext highlighter-rouge">readxl</code> package. There are other ways, but this is our preferred method because the output will be the same as the output of <code class="language-plaintext highlighter-rouge">read_csv</code>.</p>
  </blockquote>
</blockquote>

<h1 id="our-first-plot">Our first plot</h1>

<h2 id="creating-our-first-plot">Creating our first plot</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>We will be using the <code class="language-plaintext highlighter-rouge">ggplot2</code> package today to make our plots. This is a very
powerful package that creates professional looking plots and is one of the
reasons people like using R so much. All plots made using the <code class="language-plaintext highlighter-rouge">ggplot2</code> package
start by calling the <code class="language-plaintext highlighter-rouge">ggplot()</code> function. So in the tab you created for the
<code class="language-plaintext highlighter-rouge">gdp_population.R</code> file, type the following:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-ggplotDataOnly-1.png" title="plot of chunk ggplotDataOnly" alt="plot of chunk ggplotDataOnly" width="612" style="display: block; margin: auto;" /></p>

<p>To run code that you’ve typed in the editor, you have a few options. Remember
that the quickest way to run the code is by pressing
<kbd>Ctrl</kbd>+<kbd>Enter</kbd> on your keyboard. This will run the line of
code that currently contains your cursor or any highlighted code.</p>

<p>When we run this code, the <strong>Plots</strong> tab will pop to the front in the lower
right corner of the RStudio screen. Right now, we just see a big grey rectangle.</p>

<p>What we’ve done is created a ggplot object and told it we will be using the data
from the <code class="language-plaintext highlighter-rouge">gapminder_1997</code> object that we’ve loaded into R. We’ve done this by
calling the <code class="language-plaintext highlighter-rouge">ggplot()</code> function with <code class="language-plaintext highlighter-rouge">gapminder_1997</code> as the <code class="language-plaintext highlighter-rouge">data</code> argument.</p>

<p>So we’ve made a plot object, now we need to start telling it what we actually
want to draw in this plot. The elements of a plot have a bunch of properties 
like an x and y position, a size, a color, etc. When creating a data visualization, 
we can map variables in our dataset to these properties, called <strong>aesthetics</strong>, in our plot. 
In ggplot, we can do this by creating an “aesthetic mapping”, which we do with the 
<code class="language-plaintext highlighter-rouge">aes()</code> function.</p>

<p>To create our plot, we need to map variables from our <code class="language-plaintext highlighter-rouge">gapminder_1997</code> object to
ggplot aesthetics using the <code class="language-plaintext highlighter-rouge">aes()</code> function. Since we have already told
<code class="language-plaintext highlighter-rouge">ggplot</code> that we are using the data in the <code class="language-plaintext highlighter-rouge">gapminder_1997</code> object, we can
access the columns of <code class="language-plaintext highlighter-rouge">gapminder_1997</code> using the object’s column names.
(Remember, R is case-sensitive, so we have to be careful to match the column
names exactly!)</p>

<p>We are interested in whether there is a relationship between GDP and life
expectancy, so let’s start by telling our plot object that we want to map our
GDP values to the x axis of our plot. We do this by adding (<code class="language-plaintext highlighter-rouge">+</code>) information to
our plot object. Add this new line to your code and run both lines by
highlighting them and pressing <kbd>Ctrl</kbd>+<kbd>Enter</kbd> on your
keyboard:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-ggplotX-1.png" title="plot of chunk ggplotX" alt="plot of chunk ggplotX" width="612" style="display: block; margin: auto;" /></p>

<p>Note that we’ve added this new function call to a second line just to make it
easier to read. To do this we make sure that the <code class="language-plaintext highlighter-rouge">+</code> is at the end of the first
line otherwise R will assume your command ends when it starts the next row. The
<code class="language-plaintext highlighter-rouge">+</code> sign indicates not only that we are adding information, but to continue on
to the next line of code.</p>

<p>Observe that our <strong>Plot</strong> window is no longer a grey square. We now see that
we’ve mapped the <code class="language-plaintext highlighter-rouge">gdpPercap</code> column to the x axis of our plot. Note that that
column name isn’t very pretty as an x-axis label, so let’s add the <code class="language-plaintext highlighter-rouge">labs()</code>
function to make a nicer label for the x axis</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-FirstPlotAddXLabel-1.png" title="plot of chunk FirstPlotAddXLabel" alt="plot of chunk FirstPlotAddXLabel" width="612" style="display: block; margin: auto;" /></p>

<p>OK. That looks better.</p>

<blockquote class="callout">
  <h2 id="quotes-vs-no-quotes-refresher">Quotes vs. No Quotes (refresher)</h2>
  <p>Notice that when we added the label value we did so by placing the values
inside quotes. This is because we are not using a value from inside our data
object - we are providing the name directly. When you need to include actual
text values in R, they will be placed inside quotes to tell them apart from
other object or variable names.</p>

  <p>The general rule is that if you want to use values from the columns of your
data object, then you supply the name of the column without quotes, but if you
want to specify a value that does not come from your data, then use quotes.</p>
</blockquote>

<blockquote class="challenge">
  <h2 id="mapping-life-expectancy-to-the-y-axis">Mapping life expectancy to the y axis</h2>
  <p class="source">Map our <code class="language-plaintext highlighter-rouge">lifeExp</code> values to the y axis and give them a nice label.</p>

  <blockquote class="solution">
    <h2 id="solution-1">Solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <p class="source"><img src="../fig/rmd-02-FirstPlotAddY-1.png" title="plot of chunk FirstPlotAddY" alt="plot of chunk FirstPlotAddY" width="612" style="display: block; margin: auto;" /></p>
  </blockquote>
</blockquote>

<p>Excellent. We’ve now told our plot object where the x and y values are coming
from and what they stand for. But we haven’t told our object how we want it to
draw the data. There are many different plot types (bar charts, scatter plots,
histograms, etc). We tell our plot object what to draw by adding a <strong>geometry</strong>
(“geom” for short) to our object. We will talk about many different geometries
today, but for our first plot, let’s draw our data using the “points” geometry
for each value in the data set. To do this, we add <code class="language-plaintext highlighter-rouge">geom_point()</code> to our plot
object:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-FirstPlotAddPoints-1.png" title="plot of chunk FirstPlotAddPoints" alt="plot of chunk FirstPlotAddPoints" width="612" style="display: block; margin: auto;" /></p>

<p>Now we’re really getting somewhere. It finally looks like a proper plot!  We can
now see a trend in the data. It looks like countries with a larger GDP tend to
have a higher life expectancy. Let’s add a title to our plot to make that
clearer. Again, we will use the <code class="language-plaintext highlighter-rouge">labs()</code> function, but this time we will use the
<code class="language-plaintext highlighter-rouge">title =</code> argument.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Do people in wealthy countries live longer?"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-FirstPlotAddTitle-1.png" title="plot of chunk FirstPlotAddTitle" alt="plot of chunk FirstPlotAddTitle" width="612" style="display: block; margin: auto;" /></p>

<p>No one can deny we’ve made a very handsome plot! But now looking at the data, we
might be curious about learning more about the points that are the extremes of
the data. We know that we have two more pieces of data in the <code class="language-plaintext highlighter-rouge">gapminder_1997</code>
object that we haven’t used yet. Maybe we are curious if the different
continents show different patterns in GDP and life expectancy. One thing we
could do is use a different color for each of the continents. To map the
continent of each point to a color, we will again use the <code class="language-plaintext highlighter-rouge">aes()</code> function:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Do people in wealthy countries live longer?"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-FirstPlotAddColor-1.png" title="plot of chunk FirstPlotAddColor" alt="plot of chunk FirstPlotAddColor" width="612" style="display: block; margin: auto;" /></p>

<p>Here we can see that in 1997 the African countries had much lower life
expectancy than many other continents. Notice that when we add a mapping for
color, ggplot automatically provided a legend for us. It took care of assigning
different colors to each of our unique values of the <code class="language-plaintext highlighter-rouge">continent</code> variable. (Note
that when we mapped the x and y values, those drew the actual axis labels, so in
a way the axes are like the legends for the x and y values). The colors that
ggplot uses are determined by the color “scale”. Each aesthetic value we can
supply (x, y, color, etc) has a corresponding scale. Let’s change the colors to
make them a bit prettier.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Do people in wealthy countries live longer?"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-FirstPlotAddColorScale-1.png" title="plot of chunk FirstPlotAddColorScale" alt="plot of chunk FirstPlotAddColorScale" width="612" style="display: block; margin: auto;" /></p>

<p>The <code class="language-plaintext highlighter-rouge">scale_color_brewer()</code> function is just one of many you can use to change
colors. There are bunch of “palettes” that are build in. You can view them all
by running <code class="language-plaintext highlighter-rouge">RColorBrewer::display.brewer.all()</code> or check out the <a href="https://colorbrewer2.org/">Color Brewer
website</a> for more info about choosing plot colors. 
Check out the bonus exercise below for even more options.</p>

<blockquote class="challenge">
  <h2 id="bonus-exercise-changing-colors">Bonus Exercise: Changing colors</h2>
  <p>There are lots of ways to change colors when using ggplot.
The <code class="language-plaintext highlighter-rouge">scale_color_brewer()</code> function is one of many you can use to change
colors. There are bunch of “palettes” that are build in. You can view them all
by running <code class="language-plaintext highlighter-rouge">RColorBrewer::display.brewer.all()</code> or check out the <a href="https://colorbrewer2.org/">Color Brewer
website</a> for more info about choosing plot colors.
There are also lots of other fun options:</p>
  <ul>
    <li><a href="https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html">Viridis</a></li>
    <li><a href="https://github.com/katiejolly/nationalparkcolors">National parks</a></li>
    <li><a href="https://github.com/johannesbjork/LaCroixColoR">LaCroix</a></li>
    <li><a href="https://github.com/karthik/wesanderson">Wes Anderson</a></li>
  </ul>

  <p class="source">Play around with different color palettes. Feel free to install another
package and choose one of those if you want. Pick your favorite!</p>

  <blockquote class="solution">
    <h2 id="solution-2">Solution</h2>
    <p>You can use RColorBrewer::display.brewer.all() to pick a color palette. 
As a bonus, you can also use one of the packages listed above. 
Here’s an example:</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#install.packages("wesanderson") # install package from GitHub</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">wesanderson</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in library(wesanderson): there is no package called 'wesanderson'
</code></pre></div>    </div>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
</span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
</span><span class="n">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
</span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
</span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
</span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Do people in wealthy countries live longer?"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
</span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
</span><span class="n">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">wes_palette</span><span class="p">(</span><span class="s1">'Cavalcanti1'</span><span class="p">))</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error source highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in wes_palette("Cavalcanti1"): could not find function "wes_palette"
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<p>Since we have the data for the population of each country, we might be curious
what effect population might have on life expectancy and GDP per capita. Do you
think larger countries will have a longer or shorter life expectancy? Let’s find
out by mapping the population of each country to the size of our points.</p>

<blockquote class="challenge">
  <h2 id="changing-point-sizes">Changing point sizes</h2>
  <p class="source">Map the population of each country to the size of our points HINT: Is size an aesthetic or a geometry? If you’re stuck, feel free to Google it, or look at the help menu.</p>

  <blockquote class="solution">
    <h2 id="solution-3">Solution</h2>
    <p>You’ll want to use the <code class="language-plaintext highlighter-rouge">aes</code> aesthetic function to change the size:</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Do people in wealthy countries live longer?"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <p class="source"><img src="../fig/rmd-02-Pop-1.png" title="plot of chunk Pop" alt="plot of chunk Pop" width="612" style="display: block; margin: auto;" /></p>
  </blockquote>
</blockquote>

<p>There doesn’t seem to be a very strong association with population size. We can
see two very large countries with relatively low GDP per capita (but since the
per capita value is already divided by the total population, there is some
problems with separating those two values). We got another legend here for size
which is nice, but the values look a bit ugly in scientific notation. Let’s
divide all the values by 1,000,000 and label our legend “Population (in
millions)”</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Do people in wealthy countries live longer?"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="o">/</span><span class="m">1000000</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Population (in millions)"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-FirstPlotAddPop-1.png" title="plot of chunk FirstPlotAddPop" alt="plot of chunk FirstPlotAddPop" width="612" style="display: block; margin: auto;" /></p>

<p>This works because you can treat the columns in the aesthetic mappings just like
any other variables and can use functions to transform or change them at plot
time rather than having to transform your data first.</p>

<p>Good work! Take a moment to appreciate what a cool plot you made with a few
lines of code. In order to fully view its beauty you can click the “Zoom” button
in the <strong>Plots</strong> tab - it will break free from the lower right corner and open
the plot in its own window.</p>

<blockquote class="challenge">
  <h2 id="bonus-changing-shapes">Bonus: Changing shapes</h2>
  <p class="source">Instead of (or in addition to) color, change the shape of the points so each continent has a different shape. (I’m not saying this is a great thing to do - it’s just for practice!) HINT: Is shape an aesthetic or a geometry? If you’re stuck, feel free to Google it, or look at the help menu.</p>

  <blockquote class="solution">
    <h2 id="solution-4">Solution</h2>
    <p>You’ll want to use the <code class="language-plaintext highlighter-rouge">aes</code> aesthetic function to change the shape:</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Do people in wealthy countries live longer?"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="o">/</span><span class="m">1000000</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Population (in millions)"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <p class="source"><img src="../fig/rmd-02-Shape-1.png" title="plot of chunk Shape" alt="plot of chunk Shape" width="612" style="display: block; margin: auto;" /></p>
  </blockquote>
</blockquote>

<p>For our first plot we added each line of code one at a time so you could see the
exact affect it had on the output. But when you start to make a bunch of plots,
we can actually combine many of these steps so you don’t have to type as much.
For example, you can collect all the <code class="language-plaintext highlighter-rouge">aes()</code> statements and all the <code class="language-plaintext highlighter-rouge">labs()</code>
together. A more condensed version of the exact same plot would look like this:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="o">/</span><span class="m">1000000</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Do people in wealthy countries live longer?"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Population (in millions)"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-FirstPlotCondensed-1.png" title="plot of chunk FirstPlotCondensed" alt="plot of chunk FirstPlotCondensed" width="612" style="display: block; margin: auto;" /></p>

<blockquote class="challenge">
  <h2 id="storing-our-plot">Storing our plot</h2>
  <p class="source">We learned about how to save things to object names in the previous lesson. 
We can do the same thing with plots! Store our final plot in an object called <code class="language-plaintext highlighter-rouge">gdp_vs_life</code>.</p>

  <blockquote class="solution">
    <h2 id="solution-5">Solution</h2>

    <div class="language-r source highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="w"> </span><span class="n">gdp_vs_life</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="o">/</span><span class="m">1000000</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">,</span><span class="w">
   </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Do people in wealthy countries live longer?"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Population (in millions)"</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>
  </blockquote>
</blockquote>

<h2 id="saving-our-first-plot">Saving our first plot</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>Let’s say you want to share your plot with friends or co-workers who aren’t running R. It’s wise to keep all the code you used to draw the plot, but sometimes you need to make a PNG or PDF version of the plot so you can share it with your PI or post it to your Instagram story.</p>

<p>To save your plot, you can use the <code class="language-plaintext highlighter-rouge">ggsave()</code> function. A few things about <code class="language-plaintext highlighter-rouge">ggsave()</code> (the good and the bad):</p>

<ol>
  <li>[The bad] By default, <code class="language-plaintext highlighter-rouge">ggsave()</code> will save the last plot you made, but this can get confusing so it’s best to create a plot object and then save the specific plot you’re interested in instead.</li>
  <li>[The neutral] The default width and height are sometimes not great options, 
but you can supply <code class="language-plaintext highlighter-rouge">width=</code> and <code class="language-plaintext highlighter-rouge">height=</code> arguments to change them (the default values are in inches).</li>
  <li>[The good] It will determine the file type based on the name you provide.</li>
</ol>

<p>Let’s save our plot (with an informative name) as a 4x6 inch png:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gdp_vs_life</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="o">/</span><span class="m">1000000</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Do people in wealthy countries live longer?"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Population (in millions)"</span><span class="p">)</span><span class="w">

</span><span class="n">ggsave</span><span class="p">(</span><span class="n">filename</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"figures/gdp_vs_life.png"</span><span class="p">,</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdp_vs_life</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="o">=</span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="o">=</span><span class="m">4</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="o">/</span><span class="m">1000000</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Do people in wealthy countries live longer?"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Population (in millions)"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-unnamed-chunk-4-1.png" title="plot of chunk unnamed-chunk-4" alt="plot of chunk unnamed-chunk-4" width="612" style="display: block; margin: auto;" /></p>

<blockquote class="challenge">
  <h2 id="understanding-common-errors">Understanding common errors</h2>
  <p>Sometimes you accidentally type things wrong and get errors. We call these mis-types “bugs”.
Let’s go through some common ones. The most important things to remember are:</p>
  <ol>
    <li>The order of parentheses, quotes, commas, and plusses matters.</li>
    <li>Sometimes you accidentally forget a plus where you need one or include one where you don’t.</li>
  </ol>

  <p>For each of the examples below, figure out what the bug is and how to fix it. 
Feel free to copy/paste into RStudio to help you figure it out.</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Bug 1</span><span class="w">
</span></code></pre></div>  </div>

  <blockquote class="source solution">
    <h2 id="solution-6">Solution</h2>

    <div class="language-r source highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="w"> </span><span class="n">gdp_vs_life</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="o">/</span><span class="m">1000000</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Life Expectancy"</span><span class="p">,</span><span class="w">
   </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Do people in wealthy countries live longer?"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Population (in millions)"</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>
  </blockquote>
</blockquote>

<h1 id="plotting-for-data-exploration">Plotting for data exploration</h1>
<p><em><a href="#contents">Back to top</a></em></p>

<p>Many datasets are much more complex than the example we used for the first plot.
How can we find meaningful patterns in complex data and create visualizations to
convey those patterns?</p>

<h2 id="importing-datasets">Importing datasets</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>In the first plot, we looked at a smaller slice of a large dataset. To gain a
better understanding of the kinds of patterns we might observe in our own data,
we will now use the full dataset, which is stored in a file called
“gapminder_data.csv”.</p>

<p>To start, we will read in the data without using the interactive RStudio file navigation.</p>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Rows: 1704 Columns: 6
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): country, continent
dbl (4): year, pop, lifeExp, gdpPercap

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="read-in-your-own-data">Read in your own data</h2>

  <p>What argument should be provided in the below code to read in the full dataset?</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">()</span><span class="w">
</span></code></pre></div>  </div>

  <blockquote class="solution">
    <h2 id="solution-7">Solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/gapminder_data.csv"</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>
  </blockquote>
</blockquote>

<p>Let’s take a look at the full dataset. We could use <code class="language-plaintext highlighter-rouge">View()</code>, the way we did for the smaller dataset, but if your data is too big, it might take too long to load. So let’s just look at it in the console. Note that if you’re working with something other than a tibble, you might have to use the <code class="language-plaintext highlighter-rouge">head()</code> function to take a look at the first few row, but when you print out a tibble it will just print out the first few lines:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w">
</span></code></pre></div></div>

<p>Notice that this dataset has an additional column <code class="language-plaintext highlighter-rouge">year</code> compared to the smaller dataset we started with.</p>

<blockquote class="challenge">
  <h2 id="predicting-ggplot-outputs">Predicting <code class="language-plaintext highlighter-rouge">ggplot</code> outputs</h2>
  <p>Now that we have the full dataset read into our R session, let’s plot the data placing our new <code class="language-plaintext highlighter-rouge">year</code> variable on the x axis and life expectancy on the y axis. We’ve provided the code below. Notice that we’ve collapsed the plotting function options and left off some of the labels so there’s not as much code to work with.
Before running the code, read through it and see if you can predict what the plot output will look like. Then run the code and check to see if you were right!</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_data</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="n">lifeExp</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="o">=</span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">geom_point</span><span class="p">()</span><span class="w">
</span></code></pre></div>  </div>

  <p><img src="../fig/rmd-02-PlotFullGapminder-1.png" title="plot of chunk PlotFullGapminder" alt="plot of chunk PlotFullGapminder" width="612" style="display: block; margin: auto;" /></p>

</blockquote>

<p>Hmm, the plot we created in the last exercise isn’t very clear. What’s going on? Since the dataset is more complex, the plotting options we used for the smaller dataset aren’t as useful for interpreting these data. Luckily, we can add additional attributes to our plots that will make patterns more apparent. For example, we can generate a different type of plot - perhaps a line plot - and assign attributes for columns where we might expect to see patterns.</p>

<blockquote class="testimonial">
  <h2 id="pro-tip">Pro-tip</h2>

  <p>Those of us that use R on a daily basis use cheat sheets to help us remember how to use various R functions. You can also find the cheat sheets in RStudio by going to the “Help” menu and selecting “Cheat Sheets”. The ones that will be most helpful in this workshop are “Data Visualization with ggplot2”, “Data Transformation with dplyr”, “R Markdown Cheat Sheet”, and “R Markdown Reference Guide”.</p>

  <p>For things that aren’t on the cheat sheets, <a href="../06-conclusion/">Google is your best friend</a>. Even expert coders use Google when they’re stuck or trying something new!</p>

</blockquote>

<p>Our plot has a lot of points in columns which makes it hard to see trends over time. A better way to view the data showing changes over time is to use a <a href="http://www.sthda.com/english/wiki/ggplot2-line-plot-quick-start-guide-r-software-and-data-visualization">line plot</a>. Let’s try changing the geom to a line and see what happens.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="w">  </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_data</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
    </span><span class="n">geom_line</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapMinderLinePlotBad-1.png" title="plot of chunk GapMinderLinePlotBad" alt="plot of chunk GapMinderLinePlotBad" width="612" style="display: block; margin: auto;" /></p>

<p>Hmm. This doesn’t look right. By setting the color value, we got a line for each continent, but we really wanted a line for each country. We need to tell ggplot that we want to connect the values for each <code class="language-plaintext highlighter-rouge">country</code> value instead. To do this, we need to use the <code class="language-plaintext highlighter-rouge">group=</code> aesthetic.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="w">  </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_data</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
    </span><span class="n">geom_line</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapMinderLinePlot-1.png" title="plot of chunk GapMinderLinePlot" alt="plot of chunk GapMinderLinePlot" width="612" style="display: block; margin: auto;" /></p>

<p>Sometimes plots like this are called “spaghetti plots” because all the lines look like a bunch of wet noodles.</p>

<blockquote class="challenge">
  <h2 id="bonus-exercise-more-line-plots">Bonus Exercise: More line plots</h2>
  <p>Now create your own line plot comparing population and life expectancy! Looking at your plot, can you guess which two countries have experienced massive change in population from 1952-2007?</p>

  <blockquote class="solution">
    <h2 id="solution-8">Solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_data</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">geom_line</span><span class="p">()</span><span class="w">
</span></code></pre></div>    </div>

    <p><img src="../fig/rmd-02-gapminderMoreLines-1.png" title="plot of chunk gapminderMoreLines" alt="plot of chunk gapminderMoreLines" width="612" style="display: block; margin: auto;" />
(China and India are the two Asian countries that have experienced massive population growth from 1952-2007.)</p>
  </blockquote>
</blockquote>

<h2 id="discrete-plots">Discrete Plots</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>So far we’ve looked at two plot types (<code class="language-plaintext highlighter-rouge">geom_point</code> and <code class="language-plaintext highlighter-rouge">geom_line</code>) which work when both the x and y values are numeric. But sometimes one of your values may be discrete (a factor or character).</p>

<p>We’ve previously used the discrete values of the <code class="language-plaintext highlighter-rouge">continent</code> column to color in our points and lines. But now let’s try moving that variable to the <code class="language-plaintext highlighter-rouge">x</code> axis. Let’s say we are curious about comparing the distribution of the life expectancy values for each of the different continents for the <code class="language-plaintext highlighter-rouge">gapminder_1997</code> data. We can do so using a box plot. Try this out yourself in the exercise below!</p>

<blockquote class="challenge">
  <h2 id="box-plots">Box plots</h2>
  <p>Using the <code class="language-plaintext highlighter-rouge">gapminder_1997</code> data, use ggplot to create a box plot with continent on the x axis and life expectancy on the y axis. You can use the examples from earlier in the lesson as a template to remember how to pass ggplot data and map aesthetics and geometries onto the plot. If you’re really stuck, feel free to use the internet as well!</p>

  <blockquote class="solution">
    <h2 id="solution-9">Solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
 </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w">
</span></code></pre></div>    </div>

    <p><img src="../fig/rmd-02-GapBox-1.png" title="plot of chunk GapBox" alt="plot of chunk GapBox" width="612" style="display: block; margin: auto;" /></p>
  </blockquote>
</blockquote>

<p>This type of visualization makes it easy to compare the range and spread of values across groups. The “middle” 50% of the data is located inside the box and outliers that are far away from the central mass of the data are drawn as points.</p>

<blockquote class="challenge">
  <h2 id="bonus-exercise-other-discrete-geoms">Bonus Exercise: Other discrete geoms</h2>
  <p>Take a look a the ggplot <a href="https://ggplot2.tidyverse.org/">cheat sheet</a>. Find all the geoms listed under “Discrete X, Continuous Y”. Try replacing <code class="language-plaintext highlighter-rouge">geom_boxplot</code> with one of these other functions.</p>

  <blockquote class="solution">
    <h2 id="example-solution">Example solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_violin</span><span class="p">()</span><span class="w">
</span></code></pre></div>    </div>

    <p><img src="../fig/rmd-02-GapViol-1.png" title="plot of chunk GapViol" alt="plot of chunk GapViol" width="612" style="display: block; margin: auto;" /></p>
  </blockquote>
</blockquote>

<h2 id="adding-to-plot-objects">Adding to plot objects</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>In <code class="language-plaintext highlighter-rouge">ggplot2</code> you can store a plot in an object name and then add layers to the base plot. We’re going to make a base plot with continent on the x axis and life expectancy on the y axis that we can add to, so we don’t have to type out as much stuff:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">base_plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w">
</span><span class="n">base_plot</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-unnamed-chunk-6-1.png" title="plot of chunk unnamed-chunk-6" alt="plot of chunk unnamed-chunk-6" width="612" style="display: block; margin: auto;" /></p>

<p>So far we’ve only been adding one geom to each plot, but each plot object can actually contain multiple layers and each layer has it’s own geom. Let’s start with a basic boxplot:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="w"> </span><span class="n">base_plot</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapViolin-1.png" title="plot of chunk GapViolin" alt="plot of chunk GapViolin" width="612" style="display: block; margin: auto;" /></p>

<p>What does <code class="language-plaintext highlighter-rouge">base_plot</code> look like now?</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">base_plot</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-unnamed-chunk-7-1.png" title="plot of chunk unnamed-chunk-7" alt="plot of chunk unnamed-chunk-7" width="612" style="display: block; margin: auto;" /></p>

<p>It still doesn’t have the boxplots because we didn’t update the <code class="language-plaintext highlighter-rouge">base_plot</code> object.</p>

<h2 id="color-vs-fill">Color vs. Fill</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>Okay, let’s change our boxplots so that the color corresponds to continent. Remeber how to do that?</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">base_plot</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-unnamed-chunk-8-1.png" title="plot of chunk unnamed-chunk-8" alt="plot of chunk unnamed-chunk-8" width="612" style="display: block; margin: auto;" /></p>

<p>Well, that didn’t get all that colorful. That’s because objects like these boxplots have two different parts that have a color: the shape outline, and the inner part of the shape. For geoms that have an inner part, you change the fill color with <code class="language-plaintext highlighter-rouge">fill=</code> rather than <code class="language-plaintext highlighter-rouge">color=</code>, so let’s try that instead:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">base_plot</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-unnamed-chunk-9-1.png" title="plot of chunk unnamed-chunk-9" alt="plot of chunk unnamed-chunk-9" width="612" style="display: block; margin: auto;" /></p>

<p>Let’s say we want to change the fill of our plots, but to all the same color. Maybe we want our boxplots to be “cadetblue”. Let’s try it out:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">base_plot</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s2">"cadetblue"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapViolinFillWrong-1.png" title="plot of chunk GapViolinFillWrong" alt="plot of chunk GapViolinFillWrong" width="612" style="display: block; margin: auto;" />
Hmm that’s not quite what we want. In this example, we placed the fill inside the <code class="language-plaintext highlighter-rouge">aes()</code> function. Because we are using an aesthetic mapping, the “scale” for the fill will assign colors to values - in this case, we only have one value: the word “cadetblue.” Instead, we can do this by explicitly setting the color aesthetic inside the <code class="language-plaintext highlighter-rouge">geom_boxplot</code> function. Because we are assigning a color directly and not using any values from our data to do so, we do not need to use the <code class="language-plaintext highlighter-rouge">aes()</code> mapping function. Let’s try it out:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">base_plot</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s2">"cadetblue"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapViolinFill-1.png" title="plot of chunk GapViolinFill" alt="plot of chunk GapViolinFill" width="612" style="display: block; margin: auto;" /></p>

<p>That’s better! R knows lots of color names. You can see the full list if you run <code class="language-plaintext highlighter-rouge">colors()</code> in the console. Since there are so many, you can randomly choose 10 if you run <code class="language-plaintext highlighter-rouge">sample(colors(), size = 10)</code>.</p>

<blockquote class="challenge">
  <h2 id="bonus-choosing-a-color">Bonus: Choosing a color</h2>
  <p>Use <code class="language-plaintext highlighter-rouge">sample(colors(), size = 10)</code> a few times until you get an interesting sounding color name and swap that out for “cadetblue” in the violin plot example.</p>

</blockquote>

<blockquote class="challenge">
  <h2 id="transparency">Transparency</h2>
  <p>Another aesthetic that can be changed is how transparent our colors/fills are. The <code class="language-plaintext highlighter-rouge">alpha</code> parameter decides how transparent to make the colors. By default, <code class="language-plaintext highlighter-rouge">alpha = 1</code>, and our colors are completely opaque. Decreasing <code class="language-plaintext highlighter-rouge">alpha</code> increases the transparency of our colors/fills. When <code class="language-plaintext highlighter-rouge">alpha = 0</code>, it is entirely transparent.  Try changing the transparency of our boxplot. (<strong>Hint:</strong> Should alpha be inside or outside <code class="language-plaintext highlighter-rouge">aes()</code>?)</p>
  <blockquote class="solution">
    <h2 id="solution-10">Solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">base_plot</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s2">"cadetblue"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <p><img src="../fig/rmd-02-GapViolinFillSoln-1.png" title="plot of chunk GapViolinFillSoln" alt="plot of chunk GapViolinFillSoln" width="612" style="display: block; margin: auto;" /></p>
  </blockquote>
</blockquote>

<h2 id="layers">Layers</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>Now let’s add a layer of points on top of our boxplot that will show us the “raw” data:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">base_plot</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapViolinPoints-1.png" title="plot of chunk GapViolinPoints" alt="plot of chunk GapViolinPoints" width="612" style="display: block; margin: auto;" /></p>

<p>We’ve drawn the points but most of them stack up on top of each other. One way to make it easier to see all the data is to change the transparency of the points. We can do this using the <code class="language-plaintext highlighter-rouge">alpha</code> argument we learned about in the exercise above. It takes a value between 0 and 1 where 0 is entirely transparent and 1 is entirely opaque. We perform this modification inside the geom that we want to change:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">base_plot</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapViolinPointsAlpha-1.png" title="plot of chunk GapViolinPointsAlpha" alt="plot of chunk GapViolinPointsAlpha" width="612" style="display: block; margin: auto;" />
Now let’s try switching the order of <code class="language-plaintext highlighter-rouge">geom_boxplot</code> and <code class="language-plaintext highlighter-rouge">geom_point</code>. What happens? Why?</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">base_plot</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapViolinJitterBehind-1.png" title="plot of chunk GapViolinJitterBehind" alt="plot of chunk GapViolinJitterBehind" width="612" style="display: block; margin: auto;" /></p>

<p>Since we plot the <code class="language-plaintext highlighter-rouge">geom_point</code> layer first, the boxplot layer is placed on top of the <code class="language-plaintext highlighter-rouge">geom_point</code> layer, so we cannot see a lot of the points.</p>

<p>Going back to having the points on top, let’s color the points by continent. If we add a color aesthetic to the plot, then both the boxplot and the points are colored by continent:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">base_plot</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapViolinJitterLayers-1.png" title="plot of chunk GapViolinJitterLayers" alt="plot of chunk GapViolinJitterLayers" width="612" style="display: block; margin: auto;" /></p>

<p>So how do we make it so that just the points are colored but not the boxplots? Each layer can have it’s own set of aesthetic mappings. So far we’ve been using <code class="language-plaintext highlighter-rouge">aes()</code> outside of the other functions. When we do this, we are setting the “default” aesthetic mappings for the plot. But we can also set the asethetics inside the specific geom that we want to change. To do that, you can place an additional <code class="language-plaintext highlighter-rouge">aes()</code> inside of that layer:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">base_plot</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">),</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapViolinJitterSpecific-1.png" title="plot of chunk GapViolinJitterSpecific" alt="plot of chunk GapViolinJitterSpecific" width="612" style="display: block; margin: auto;" /></p>

<p>Nice! Both <code class="language-plaintext highlighter-rouge">geom_boxplot</code> and <code class="language-plaintext highlighter-rouge">geom_point</code> will inherit the default values of <code class="language-plaintext highlighter-rouge">aes(continent, lifeExp)</code> in the base plot, but only <code class="language-plaintext highlighter-rouge">geom_point</code> will also use <code class="language-plaintext highlighter-rouge">aes(color = continent)</code>.</p>

<blockquote class="callout">
  <h2 id="aesthetics-inside-the-ggplot-function">Aesthetics inside the <code class="language-plaintext highlighter-rouge">ggplot()</code> function</h2>
  <p>We can also map our default aesthetics by passing the values to the <code class="language-plaintext highlighter-rouge">ggplot()</code> function call as is sometimes more common:</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">,</span><span class="w"> </span><span class="n">mapping</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w">
</span></code></pre></div>  </div>

  <p><img src="../fig/rmd-02-unnamed-chunk-10-1.png" title="plot of chunk unnamed-chunk-10" alt="plot of chunk unnamed-chunk-10" width="612" style="display: block; margin: auto;" /></p>
</blockquote>

<h2 id="univariate-plots">Univariate Plots</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>We jumped right into make plots with multiple columns. But what if we wanted to take a look at just one column? In that case, we only need to specify a mapping for <code class="language-plaintext highlighter-rouge">x</code> and choose an appropriate geom.</p>

<h3 id="univariate-continuous">Univariate continuous</h3>

<p>Let’s start with a <a href="https://www.thoughtco.com/what-is-a-histogram-3126359">histogram</a> to see the range and spread of the life expectancy values</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_histogram</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
</code></pre></div></div>

<p><img src="../fig/rmd-02-GapLifeHist-1.png" title="plot of chunk GapLifeHist" alt="plot of chunk GapLifeHist" width="612" style="display: block; margin: auto;" /></p>

<p>You should not only see the plot in the plot window, but also a message telling you to choose a better bin value. Histograms can look very different depending on the number of bars you decide to draw. The default is 30. Let’s try setting a different value by explicitly passing a <code class="language-plaintext highlighter-rouge">bin=</code> argument to the <code class="language-plaintext highlighter-rouge">geom_histogram</code> later.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_histogram</span><span class="p">(</span><span class="n">bins</span><span class="o">=</span><span class="m">20</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapLifeHistBins-1.png" title="plot of chunk GapLifeHistBins" alt="plot of chunk GapLifeHistBins" width="612" style="display: block; margin: auto;" /></p>

<p>Try different values like 5 or 50 to see how the plot changes.</p>

<blockquote class="challenge">
  <h2 id="bonus-exercise-one-variable-plots">Bonus Exercise: One variable plots</h2>
  <p>Rather than a histogram, choose one of the other geometries listed under “One Variable” continuous plots on the ggplot <a href="https://ggplot2.tidyverse.org/">cheat sheet</a>.</p>

  <blockquote class="solution">
    <h2 id="example-solution-1">Example solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_density</span><span class="p">()</span><span class="w">
</span></code></pre></div>    </div>

    <p><img src="../fig/rmd-02-GapLifeDens1-1.png" title="plot of chunk GapLifeDens1" alt="plot of chunk GapLifeDens1" width="612" style="display: block; margin: auto;" /></p>
  </blockquote>
</blockquote>

<h3 id="univariate-discrete">Univariate discrete</h3>

<p>What if we want to plot a univariate discrete variable, like continent? For this, we can use a bar chart.</p>

<blockquote class="challenge">
  <h2 id="exercise-discrete-univariate-plots">Exercise: Discrete univariate plots</h2>
  <p>Create a bar plot of <code class="language-plaintext highlighter-rouge">continent</code>. You can try guessing the geom or look it up on the cheat sheet or Internet.</p>

  <blockquote class="solution">
    <h2 id="example-solution-2">Example solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_bar</span><span class="p">()</span><span class="w">
</span></code></pre></div>    </div>

    <p><img src="../fig/rmd-02-barplot-1.png" title="plot of chunk barplot" alt="plot of chunk barplot" width="612" style="display: block; margin: auto;" /></p>
  </blockquote>
</blockquote>

<h2 id="plot-themes">Plot Themes</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>Our plots are looking pretty nice, but what’s with that grey background? While you can change various elements of a <code class="language-plaintext highlighter-rouge">ggplot</code> object manually (background color, grid lines, etc.) the <code class="language-plaintext highlighter-rouge">ggplot</code> package also has a bunch of nice built-in themes to change the look of your graph. For example, let’s try adding <code class="language-plaintext highlighter-rouge">theme_bw()</code> to our histogram:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_histogram</span><span class="p">(</span><span class="n">bins</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme_bw</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapLifeHistBinsClassicTheme-1.png" title="plot of chunk GapLifeHistBinsClassicTheme" alt="plot of chunk GapLifeHistBinsClassicTheme" width="612" style="display: block; margin: auto;" /></p>

<p>Try out a few other themes, to see which you like: <code class="language-plaintext highlighter-rouge">theme_classic()</code>, <code class="language-plaintext highlighter-rouge">theme_linedraw()</code>, <code class="language-plaintext highlighter-rouge">theme_minimal()</code>.</p>

<blockquote class="challenge">
  <h2 id="rotating-x-axis-labels">Rotating x axis labels</h2>
  <p>Often, you’ll want to change something about the theme that you don’t know how to do off the top of your head. When this happens, you can do an Internet search to help find what you’re looking for. To practice this, search the Internet to figure out how to rotate the x axis labels 90 degrees. Then try it out using the histogram plot we made above.</p>

  <blockquote class="solution">
    <h2 id="solution-11">Solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_histogram</span><span class="p">(</span><span class="n">bins</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">90</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="o">=</span><span class="m">1</span><span class="p">))</span><span class="w">
</span></code></pre></div>    </div>

    <p><img src="../fig/rmd-02-GapLifeDens2-1.png" title="plot of chunk GapLifeDens2" alt="plot of chunk GapLifeDens2" width="612" style="display: block; margin: auto;" /></p>
  </blockquote>
</blockquote>

<h2 id="facets">Facets</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>If you have a lot of different columns to try to plot or have distinguishable subgroups in your data, a powerful plotting technique called faceting might come in handy. When you facet your plot, you basically make a bunch of smaller plots and combine them together into a single image. Luckily, <code class="language-plaintext highlighter-rouge">ggplot</code> makes this very easy. Let’s start with a simplified version of our first plot</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapNoFacet-1.png" title="plot of chunk GapNoFacet" alt="plot of chunk GapNoFacet" width="612" style="display: block; margin: auto;" /></p>

<p>The first time we made this plot, we colored the points differently for each of the continents. This time let’s actually draw a separate box for each continent. We can do this with <code class="language-plaintext highlighter-rouge">facet_wrap()</code></p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_wrap</span><span class="p">(</span><span class="n">vars</span><span class="p">(</span><span class="n">continent</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapFacetWrap-1.png" title="plot of chunk GapFacetWrap" alt="plot of chunk GapFacetWrap" width="612" style="display: block; margin: auto;" /></p>

<p>Note that <code class="language-plaintext highlighter-rouge">facet_wrap</code> requires this extra helper function called <code class="language-plaintext highlighter-rouge">vars()</code> in order to pass in the column names. It’s a lot like the <code class="language-plaintext highlighter-rouge">aes()</code> function, but it doesn’t require an aesthetic name. We can see in this output that we get a separate box with a label for each continent so that only the points for that continent are in that box.</p>

<p>The other faceting function ggplot provides is <code class="language-plaintext highlighter-rouge">facet_grid()</code>. The main difference is that <code class="language-plaintext highlighter-rouge">facet_grid()</code> will make sure all of your smaller boxes share a common axis. In this example, we will stack all the boxes on top of each other into rows so that their x axes all line up.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder_1997</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_grid</span><span class="p">(</span><span class="n">rows</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">vars</span><span class="p">(</span><span class="n">continent</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<p><img src="../fig/rmd-02-GapFacetGrid-1.png" title="plot of chunk GapFacetGrid" alt="plot of chunk GapFacetGrid" width="612" style="display: block; margin: auto;" /></p>

<p>Unlike the <code class="language-plaintext highlighter-rouge">facet_wrap</code> output where each box got its own x and y axis, with <code class="language-plaintext highlighter-rouge">facet_grid()</code>, there is only one x axis along the bottom.</p>

<h1 id="creating-and-saving-your-own-plot">Creating and saving your own plot</h1>
<p><em><a href="#contents">Back to top</a></em></p>

<blockquote class="challenge">
  <h2 id="bonus-exercise-create-and-save-a-plot">Bonus Exercise: Create and save a plot</h2>
  <p>Create your own plot using <code class="language-plaintext highlighter-rouge">ggplot()</code>, store it in an object named <code class="language-plaintext highlighter-rouge">my_plot</code>, and save the plot using <code class="language-plaintext highlighter-rouge">ggsave()</code>.</p>

  <blockquote class="solution">
    <h2 id="example-solution-3">Example solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">my_plot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="p">)</span><span class="o">+</span><span class="w">
  </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">continent</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"orange"</span><span class="p">)</span><span class="o">+</span><span class="w">
  </span><span class="n">theme_bw</span><span class="p">()</span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Continent"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GDP Per Capita"</span><span class="p">)</span><span class="w">

</span><span class="n">ggsave</span><span class="p">(</span><span class="s2">"my_awesome_plot.jpg"</span><span class="p">,</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">my_plot</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="o">=</span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="o">=</span><span class="m">4</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>
  </blockquote>
</blockquote>

<h1 id="applying-it-to-your-own-data">Applying it to your own data</h1>
<p><em><a href="#contents">Back to top</a></em></p>

<p>Now that we’ve learned how impactful effective data visualization can be, and how to create informative visuals in R, it’s time for you to start thinking about what you want to do with your own data!</p>

<p>Pair up with another learner and discuss your data and what type of exploratory data analysis you would like to perform.</p>

<p>Questions to conisder:</p>

<ol>
  <li>What information do you have in your data set?</li>
  <li>Are there specific questions are you interested in investigating with the data you have?</li>
  <li>What visuals can you create to explore your data? Draw some out on paper! What will these plots show you?</li>
  <li>Is your data in a format that can easily be read into R? How can you improve the formatting of the contents of the file?</li>
</ol>

<h1 id="glossary-of-terms">Glossary of terms</h1>
<p><em><a href="#contents">Back to top</a></em></p>

<ul>
  <li>Tibble: the way tabular data is stored in R when using the tidyverse. We may also call it a data frame.</li>
  <li>Aesthetic: a visual property of the objects (geoms) drawn in your plot (like x position, y position, color, size, etc)</li>
  <li>Aesthetic mapping (aes): This is how we connect a visual property of the plot to a column of our data</li>
  <li>Comments: lines of text in our code after a <code class="language-plaintext highlighter-rouge">#</code> that are ignored (not evaluated) by R</li>
  <li>Geometry (geom): this describes the things that are actually drawn on the plot (like points or lines)</li>
  <li>Facets: Dividing your data into non-overlapping groups and making a small plot for each subgroup</li>
  <li>Layer: Each ggplot is made up of one or more layers. Each layer contains one geometry and may also contain custom aesthetic mappings and private data</li>
  <li>Factor: a way of storing data to let R know the values are discrete so they get special treatment</li>
</ul>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>Use <code class="language-plaintext highlighter-rouge">read_csv()</code> to read tabular data in R.</p>
</li>
    
    <li><p>Geometries are the visual elements drawn on data visualizations (lines, points, etc.), and aesthetics are the visual properties of those geometries that are assigned to variables in the data (color, position, etc.).</p>
</li>
    
    <li><p>Use <code class="language-plaintext highlighter-rouge">ggplot()</code> and geoms to create data visualizations, and save them using <code class="language-plaintext highlighter-rouge">ggsave()</code>.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="r-for-data-analysis" class="maintitle">R for Data Analysis</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 75 min
      <br />
      <strong>Exercises:</strong> 15 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How can I summarize my data in R?</p>
</li>
	
	<li><p>How can R help make my research more reproducible?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>To become familiar with the functions of the <code class="language-plaintext highlighter-rouge">dplyr</code> and <code class="language-plaintext highlighter-rouge">tidyr</code> packages.</p>
</li>
	
	<li><p>To be able to create plots and summary tables to answer analysis questions.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<h3 id="contents">Contents</h3>

<ol>
  <li><a href="#plotting-review">Plotting review</a></li>
  <li><a href="#getting-started">Getting started</a>
    <ul>
      <li><a href="#loading-in-the-data">Loading in the data</a></li>
    </ul>
  </li>
  <li><a href="#intro-data-analysis">An introduction to data analysis in R using <code class="language-plaintext highlighter-rouge">dplyr</code></a>
    <ul>
      <li><a href="#get-stats-fast-with-summarize">Get stats fast with <code class="language-plaintext highlighter-rouge">summarize()</code></a></li>
      <li><a href="#narrow-down-rows-with-filter">Narrow down rows with <code class="language-plaintext highlighter-rouge">filter()</code></a></li>
      <li><a href="#grouping-rows-using-group_by">Grouping rows using <code class="language-plaintext highlighter-rouge">group_by()</code></a></li>
      <li><a href="#make-new-variables-with-mutate">Make new variables with <code class="language-plaintext highlighter-rouge">mutate()</code></a></li>
      <li><a href="#subset-columns-using-select">Subset columns using <code class="language-plaintext highlighter-rouge">select()</code></a></li>
      <li><a href="#changing-the-shape-of-the-data">Changing the shape of the data</a></li>
    </ul>
  </li>
  <li><a href="#applying-it-to-your-own-data">Applying it to your own data</a></li>
</ol>

<h1 id="plotting-review">Plotting review</h1>

<p>First, navigate to the un-reports directory however you’d like and open <code class="language-plaintext highlighter-rouge">un-report.Rproj</code>.
This should open the un-report R project in RStudio.
You can check this by seeing if the Files in the bottom right of RStudio are the ones in your <code class="language-plaintext highlighter-rouge">un-report</code> directory.</p>

<p>The very first step is to read in the gapminder dataset, so do that first! Also load the <code class="language-plaintext highlighter-rouge">tidyverse</code> package.</p>

<blockquote class="solution">
  <h2 id="solution">Solution</h2>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span><span class="w">
</span></code></pre></div>  </div>

  <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──
</code></pre></div>  </div>

  <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>✔ ggplot2 3.3.6     ✔ purrr   0.3.4
✔ tibble  3.1.7     ✔ dplyr   1.0.9
✔ tidyr   1.2.0     ✔ stringr 1.4.0
✔ readr   2.1.2     ✔ forcats 0.5.1
</code></pre></div>  </div>

  <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
</code></pre></div>  </div>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s1">'data/gapminder_data.csv'</span><span class="p">)</span><span class="w">
</span></code></pre></div>  </div>

  <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Rows: 1704 Columns: 6
</code></pre></div>  </div>

  <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): country, continent
dbl (4): year, pop, lifeExp, gdpPercap

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
</code></pre></div>  </div>
</blockquote>
<h4 id="investigating-population-over-time">Investigating population over time.</h4>
<p><em><a href="#contents">Back to top</a></em></p>

<p>Next, make a scatter plot of year vs. population, separated into a plot for each contient. Feel free to look back at the content from yesterday if you want!</p>

<blockquote class="solution">
  <h2 id="solution-1">Solution</h2>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">year</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="n">pop</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
</span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
</span><span class="n">facet_wrap</span><span class="p">(</span><span class="n">vars</span><span class="p">(</span><span class="n">continent</span><span class="p">))</span><span class="w">
</span></code></pre></div>  </div>

  <p><img src="../fig/rmd-03-unnamed-chunk-3-1.png" title="plot of chunk unnamed-chunk-3" alt="plot of chunk unnamed-chunk-3" width="612" style="display: block; margin: auto;" /></p>
</blockquote>

<p>It seems like there are a couple outliers. By the end of today, we’ll be able to identify what countries those are and how to exclude them from the plot.</p>

<h1 id="getting-started">Getting Started</h1>
<p><em><a href="#contents">Back to top</a></em></p>

<p>Yesterday we spent a lot of time making plots in R using the ggplot2 package. Visualizing data using plots is a very powerful skill in R, but what if we would like to work with only a subset of our data? Or clean up messy data, calculate summary statistics, create a new variable, or join two datasets together? There are several different methods for doing this in R, and we will touch on a few today using functions the <code class="language-plaintext highlighter-rouge">dplyr</code> package.</p>

<p>First, we will create a new R Script file for our work. Open RStudio. Choose “File” &gt; “New File” &gt; “RScript”. We will save this file as <code class="language-plaintext highlighter-rouge">un_data_analysis.R</code></p>

<h3 id="loading-in-the-data">Loading in the data</h3>

<p>We will start by importing the complete gapminder dataset that we used yesterday into our fresh new R session. Yesterday we did this using a “point-and-click” commands. Today let’s type them into the console ourselves: <code class="language-plaintext highlighter-rouge">gapminder_data &lt;- read_csv("data/gapminder_data.csv")</code></p>

<blockquote class="challenge">
  <h2 id="exercise">Exercise</h2>

  <p>If we look in the console now, we’ll see we’ve received an error message saying that R “could not find the function <code class="language-plaintext highlighter-rouge">read_csv()</code>”. <em>Hint: Packages…</em></p>

  <blockquote class="solution">
    <h2 id="solution-2">Solution</h2>

    <p>What this means is that R cannot find the function we are trying to call. The reason for this usually is that we are trying to run a function from a package that we have not yet loaded. This is a very common error message that you will probably see a lot when using R. It’s important to remember that you will need to load any packages you want to use into R each time you start a new session. The <code class="language-plaintext highlighter-rouge">read_csv</code> function comes from the <code class="language-plaintext highlighter-rouge">readr</code> package which is included in the <code class="language-plaintext highlighter-rouge">tidyverse</code> package so we will just load the <code class="language-plaintext highlighter-rouge">tidyverse</code> package and run the import code again.</p>
  </blockquote>
</blockquote>

<p>Now that we know what’s wrong, We will use the <code class="language-plaintext highlighter-rouge">read_csv()</code> function from the Tidyverse <code class="language-plaintext highlighter-rouge">readr</code> package. Load the <code class="language-plaintext highlighter-rouge">tidyverse</code> package and gapminder dataset using the code below.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>The output in your console shows that by doing this, we attach several useful packages for data wrangling, including <code class="language-plaintext highlighter-rouge">readr</code>. Check out these packages and their documentation at <a href="https://www.tidyverse.org">tidyverse.org</a></p>

<blockquote>
  <p><strong>Reminder:</strong> Many of these packages, including <code class="language-plaintext highlighter-rouge">dplyr</code> , come with “Cheatsheets” found under the <strong>Help</strong> RStudio menu tab.</p>
</blockquote>

<p>Reload your data:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/gapminder_data.csv"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Rows: 1704 Columns: 6
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): country, continent
dbl (4): year, pop, lifeExp, gdpPercap

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
</code></pre></div></div>

<p>Notice that the output of the <code class="language-plaintext highlighter-rouge">read_csv()</code> function is pretty informative. It tells us the name of all of our column headers as well as how it interpreted the data type. This birds-eye-view can help you take a quick look that everything is how we expect it to be.</p>

<p>Now we have the tools necessary to work through this lesson.</p>

<h1 id="intro-data-analysis">An introduction to data analysis in R using <code class="language-plaintext highlighter-rouge">dplyr</code></h1>

<h2 id="get-stats-fast-with-summarize">Get stats fast with <code class="language-plaintext highlighter-rouge">summarize()</code></h2>

<p><a href="#contents"><em>Back to top</em></a></p>

<p>Let’s say we would like to know what is the mean (average) life expecteny in the dataset. R has a built in function function called <code class="language-plaintext highlighter-rouge">mean()</code> that will calculate this value for us. We can apply that function to our lifeExp column using the <code class="language-plaintext highlighter-rouge">summarize()</code> function. Here’s what that looks like:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">summarize</span><span class="p">(</span><span class="n">gapminder_data</span><span class="p">,</span><span class="w"> </span><span class="n">averageLifeExp</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">lifeExp</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1 × 1
  averageLifeExp
           &lt;dbl&gt;
1           59.5
</code></pre></div></div>

<p>When we call <code class="language-plaintext highlighter-rouge">summarize()</code>, we can use any of the column names of our data object as values to pass to other functions. <code class="language-plaintext highlighter-rouge">summarize()</code> will return a new data object and our value will be returned as a column.</p>

<blockquote>
  <p><strong>Note:</strong> The <code class="language-plaintext highlighter-rouge">summarize()</code> and <code class="language-plaintext highlighter-rouge">summarise()</code> perform identical functions.</p>
</blockquote>

<p>We name this new column so we can use in a future argument. So the <code class="language-plaintext highlighter-rouge">averageLifeExp=</code> part tells <code class="language-plaintext highlighter-rouge">summarize()</code> to use “averageLifeExp” as the name of the new column. Note that you don’t have to quotes around this new name as long as it starts with a letter and doesn’t include a space.</p>

<p>Instead of including the data as an argument, we can use the <em>pipe operator</em> <code class="language-plaintext highlighter-rouge">%&gt;%</code> to pass the data value into the <code class="language-plaintext highlighter-rouge">summarize</code> function.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="n">summarize</span><span class="p">(</span><span class="n">averageLifeExp</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">lifeExp</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1 × 1
  averageLifeExp
           &lt;dbl&gt;
1           59.5
</code></pre></div></div>

<p>This line of code will do the exact same thing as our first summary command, but the piping function tells R to use the <code class="language-plaintext highlighter-rouge">gapminder_data</code> dataframe as the first argument in the next function.</p>

<p>This lets us “chain” together multiple functions, which will be helpful later. Note that the pipe (<code class="language-plaintext highlighter-rouge">%&gt;%</code>) is a bit different from using the ggplot plus (<code class="language-plaintext highlighter-rouge">+</code>). Pipes take the output from the left side and use it as input to the right side. Plusses layer on additional information (right side) to a preexisting plot (left side).</p>

<p>We can also add an <kdb>Enter</kdb> to make it look nicer:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarize</span><span class="p">(</span><span class="n">averageLifeExp</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">lifeExp</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1 × 1
  averageLifeExp
           &lt;dbl&gt;
1           59.5
</code></pre></div></div>

<p>Using the <em>pipe operator</em> <code class="language-plaintext highlighter-rouge">%&gt;%</code> and enter command makes our code more readable. The  <em>pipe operator</em> <code class="language-plaintext highlighter-rouge">%&gt;%</code> also helps to avoid using nested function and minimizes the need for new variables.</p>

<p>Since we use the pipe operator so often, there is a keyboard shortcut for it in RStudio. You can press <kdb>Ctrl</kdb>+<kdb>Shift</kdb>+<kdb>M<kdb> on Windows or <kdb>Cmd<kdb>+<kdb>Shift<kdb>+<kdb>M<kdb> on a Mac.</kdb></kdb></kdb></kdb></kdb></kdb></kdb></kdb></p>

<blockquote class="callout">
  <h2 id="pro-tip-saving-a-new-dataframe">Pro tip: Saving a new dataframe</h2>
  <p>Notice that when we run the following code, we are not actually saving a new variable:</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarize</span><span class="p">(</span><span class="n">averageLifeExp</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">lifeExp</span><span class="p">))</span><span class="w">
</span></code></pre></div>  </div>
  <p>This simply outputs what we have created, but does not change actually change <code class="language-plaintext highlighter-rouge">gapminder_data</code> or save a new dataframe. To save a new dataframe, we could run:</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data_summarized</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarize</span><span class="p">(</span><span class="n">averageLifeExp</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">lifeExp</span><span class="p">))</span><span class="w">
</span></code></pre></div>  </div>
  <p>Or if we want to change <code class="language-plaintext highlighter-rouge">gapminder_data</code> itself:</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarize</span><span class="p">(</span><span class="n">averageLifeExp</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">lifeExp</span><span class="p">))</span><span class="w">
</span></code></pre></div>  </div>
  <p>IMPORTANT: This would overwrite the existing <code class="language-plaintext highlighter-rouge">gapminder_data</code> object.</p>

  <p>For now, we will not be saving dataframes, since we are just experimenting with <code class="language-plaintext highlighter-rouge">dplyr</code> functions, but it will be useful later on in this lesson.</p>

</blockquote>

<h2 id="narrow-down-rows-with-filter">Narrow down rows with <code class="language-plaintext highlighter-rouge">filter()</code></h2>

<p><a href="#contents"><em>Back to top</em></a></p>

<p>Let’s take a look at the value we just calculated, which tells us the average life expectancy for all rows in the data was 59.5. That seems a bit low, doesn’t it? What’s going on?</p>

<p>Well, remember the dataset contains rows from many different years and many different countries. It’s likely that life expectancy has increased overtime, so it may not make sense to average over all the years at the same time.</p>

<p>Use <code class="language-plaintext highlighter-rouge">summarize()</code> to find the most recent year in the data set. We can use the <code class="language-plaintext highlighter-rouge">max()</code> function to return the maximum value.</p>

<blockquote class="challenge">
  <h2 id="practice-using-the--to-summarize-data">Practice using the <code class="language-plaintext highlighter-rouge">%&gt;%</code> to summarize data</h2>

  <p>Find the most recent year using the piping function.</p>

  <blockquote class="solution">
    <h2 id="solution-3">Solution:</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">  
 </span><span class="n">summarize</span><span class="p">(</span><span class="n">recent_year</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">max</span><span class="p">(</span><span class="n">year</span><span class="p">))</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1 × 1
  recent_year
        &lt;dbl&gt;
1        2007
</code></pre></div>    </div>

  </blockquote>
</blockquote>

<p>So we see that the most recent year in the dataset is 2007. Let’s calculate the life expectancy for all countries for only that year. To do that, we will use the <code class="language-plaintext highlighter-rouge">filter()</code> function to only use rows for that year before calculating the mean value.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">filter</span><span class="p">(</span><span class="n">year</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">2007</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarize</span><span class="p">(</span><span class="n">average</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">lifeExp</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1 × 1
  average
    &lt;dbl&gt;
1    67.0
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="filtering-the-dataset">Filtering the dataset</h2>

  <p>What is the average GDP per capita for the first year in the dataset? <em>Hint: the column headers identified by <code class="language-plaintext highlighter-rouge">read_csv()</code> showed us there was a column called gdpPercap in the dataset</em></p>

  <blockquote class="solution">
    <h2 id="solution-4">Solution</h2>

    <p>Identify the earliest year in our dataset using <code class="language-plaintext highlighter-rouge">min()</code> and <code class="language-plaintext highlighter-rouge">summarize()</code></p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
</span><span class="n">summarize</span><span class="p">(</span><span class="n">first_year</span><span class="o">=</span><span class="nf">min</span><span class="p">(</span><span class="n">year</span><span class="p">))</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1 × 1
  first_year
       &lt;dbl&gt;
1       1952
</code></pre></div>    </div>

    <p>We see here that the first year in the dataset is 1952. Filter to only 1952, and determin the average GDP per capita.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
</span><span class="n">filter</span><span class="p">(</span><span class="n">year</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1952</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
</span><span class="n">summarize</span><span class="p">(</span><span class="n">average_gdp</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">gdpPercap</span><span class="p">))</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output source highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1 × 1
  average_gdp
        &lt;dbl&gt;
1       3725.
</code></pre></div>    </div>
    <p>By combining <code class="language-plaintext highlighter-rouge">filter()</code> and <code class="language-plaintext highlighter-rouge">summarize()</code> we were able to calculate the mean GDP per capita in the year 1952.</p>
  </blockquote>
</blockquote>

<p>Notice how the pipe operator (<code class="language-plaintext highlighter-rouge">%&gt;%</code>) allows us to combine these two simple steps into a more complicated data extraction?. We took the data, filtered out the rows, then took the mean value. The argument we pass to <code class="language-plaintext highlighter-rouge">filter()</code> needs to be some expression that will return TRUE or FALSE. We can use comparisons like <code class="language-plaintext highlighter-rouge">&gt;</code> (greater than) and <code class="language-plaintext highlighter-rouge">&lt;</code> (less than) for example. Here we tested for equality using a double equals sign <code class="language-plaintext highlighter-rouge">==</code>. You use <code class="language-plaintext highlighter-rouge">==</code> (double equals) when testing if two values are equal, and you use <code class="language-plaintext highlighter-rouge">=</code> (single equals) when naming arguments that you are passing to functions. Try changing it to use <code class="language-plaintext highlighter-rouge">filter(year = 2007)</code> and see what happens.</p>

<h2 id="grouping-rows-using-group_by">Grouping rows using <code class="language-plaintext highlighter-rouge">group_by()</code></h2>

<p><a href="#contents"><em>Back to top</em></a></p>

<p>We see that the life expectancy in 2007 is much larger than the value we got using all of the rows. It seems life expectancy is increasing which is good news. But now we might be interested in calculating the average for each year. Rather that doing a bunch of different <code class="language-plaintext highlighter-rouge">filter()</code> statements, we can instead use the <code class="language-plaintext highlighter-rouge">group_by()</code> function. The function allows us to tell the code to treat the rows in logical groups, so rather than summarizing over all the rows, we will get one summary value for each group. Here’s what that will look like:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">group_by</span><span class="p">(</span><span class="n">year</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarize</span><span class="p">(</span><span class="n">average</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">lifeExp</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 12 × 2
    year average
   &lt;dbl&gt;   &lt;dbl&gt;
 1  1952    49.1
 2  1957    51.5
 3  1962    53.6
 4  1967    55.7
 5  1972    57.6
 6  1977    59.6
 7  1982    61.5
 8  1987    63.2
 9  1992    64.2
10  1997    65.0
11  2002    65.7
12  2007    67.0
</code></pre></div></div>

<p>The <code class="language-plaintext highlighter-rouge">group_by()</code> function expects you to pass in the name of a column (or multiple columns separated by comma) in your data.</p>

<p>Note that you might get a message about the summarize function regrouping the output by ‘year’. This simply indicates what the function is grouping by.</p>

<blockquote class="challenge">
  <h2 id="grouping-the-data">Grouping the data</h2>

  <p>Try calculating the average life expectancy by continent.</p>

  <blockquote class="solution">
    <h2 id="solution-5">Solution</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
</span><span class="n">group_by</span><span class="p">(</span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
</span><span class="n">summarize</span><span class="p">(</span><span class="n">average</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">lifeExp</span><span class="p">))</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output source highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 5 × 2
  continent average
  &lt;chr&gt;       &lt;dbl&gt;
1 Africa       48.9
2 Americas     64.7
3 Asia         60.1
4 Europe       71.9
5 Oceania      74.3
</code></pre></div>    </div>

    <p>By combining <code class="language-plaintext highlighter-rouge">group_by()</code> and <code class="language-plaintext highlighter-rouge">summarize()</code> we are able to calculate the mean life expectancy by continent.</p>
  </blockquote>
</blockquote>

<p>You can also create more than one new column when you call <code class="language-plaintext highlighter-rouge">summarize()</code>. To do so, you must separate your columns with a comma. Building on the code from the last exercise, let’s add a new column that calculates the minimum life expectancy for each continent.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">group_by</span><span class="p">(</span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarize</span><span class="p">(</span><span class="n">average</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">lifeExp</span><span class="p">),</span><span class="w"> </span><span class="n">min</span><span class="o">=</span><span class="nf">min</span><span class="p">(</span><span class="n">lifeExp</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 5 × 3
  continent average   min
  &lt;chr&gt;       &lt;dbl&gt; &lt;dbl&gt;
1 Africa       48.9  23.6
2 Americas     64.7  37.6
3 Asia         60.1  28.8
4 Europe       71.9  43.6
5 Oceania      74.3  69.1
</code></pre></div></div>

<h2 id="make-new-variables-with-mutate">Make new variables with <code class="language-plaintext highlighter-rouge">mutate()</code></h2>

<p><a href="#contents"><em>Back to top</em></a></p>

<p>Each time we ran <code class="language-plaintext highlighter-rouge">summarize()</code>, we got back fewer rows than passed in. We either got one row back, or one row per group. But sometimes we want to create a new column in our data without changing the number of rows. The function we use to create new columns is called <code class="language-plaintext highlighter-rouge">mutate()</code>.</p>

<p>We have a column for the population and the GDP per capita. If we wanted to get the total GDP, we could multiply the per capita GDP values by the total population. Here’s what such a <code class="language-plaintext highlighter-rouge">mutate()</code> command would look like:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">gdp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1,704 × 7
   country      year      pop continent lifeExp gdpPercap          gdp
   &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;       &lt;dbl&gt;     &lt;dbl&gt;        &lt;dbl&gt;
 1 Afghanistan  1952  8425333 Asia         28.8      779.  6567086330.
 2 Afghanistan  1957  9240934 Asia         30.3      821.  7585448670.
 3 Afghanistan  1962 10267083 Asia         32.0      853.  8758855797.
 4 Afghanistan  1967 11537966 Asia         34.0      836.  9648014150.
 5 Afghanistan  1972 13079460 Asia         36.1      740.  9678553274.
 6 Afghanistan  1977 14880372 Asia         38.4      786. 11697659231.
 7 Afghanistan  1982 12881816 Asia         39.9      978. 12598563401.
 8 Afghanistan  1987 13867957 Asia         40.8      852. 11820990309.
 9 Afghanistan  1992 16317921 Asia         41.7      649. 10595901589.
10 Afghanistan  1997 22227415 Asia         41.8      635. 14121995875.
# … with 1,694 more rows
</code></pre></div></div>

<p>This will add a new column called “gdp” to our data. We use the column names as if they were regular values that we want to perform mathematical operations on and provide the name in front of an equals sign like we have done with <code class="language-plaintext highlighter-rouge">summarize()</code></p>

<blockquote class="challenge">
  <h2 id="mutate"><code class="language-plaintext highlighter-rouge">mutate()</code></h2>
  <p>We can also multiply by constants or other numbers using mutate - remember how in the plotting lesson we made a plot with population in millions? Try making a new column for this dataframe called popInMillions that is the population in million.</p>

  <blockquote class="solution">
    <h2 id="solution-6">Solution:</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">  
</span><span class="n">mutate</span><span class="p">(</span><span class="n">gdp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">gdpPercap</span><span class="p">,</span><span class="w"> </span><span class="n">popInMillions</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="m">1000000</span><span class="p">)</span><span class="w">  
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1,704 × 8
   country      year      pop continent lifeExp gdpPercap      gdp popInMillions
   &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;       &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;         &lt;dbl&gt;
 1 Afghanistan  1952  8425333 Asia         28.8      779.  6.57e 9          8.43
 2 Afghanistan  1957  9240934 Asia         30.3      821.  7.59e 9          9.24
 3 Afghanistan  1962 10267083 Asia         32.0      853.  8.76e 9         10.3 
 4 Afghanistan  1967 11537966 Asia         34.0      836.  9.65e 9         11.5 
 5 Afghanistan  1972 13079460 Asia         36.1      740.  9.68e 9         13.1 
 6 Afghanistan  1977 14880372 Asia         38.4      786.  1.17e10         14.9 
 7 Afghanistan  1982 12881816 Asia         39.9      978.  1.26e10         12.9 
 8 Afghanistan  1987 13867957 Asia         40.8      852.  1.18e10         13.9 
 9 Afghanistan  1992 16317921 Asia         41.7      649.  1.06e10         16.3 
10 Afghanistan  1997 22227415 Asia         41.8      635.  1.41e10         22.2 
# … with 1,694 more rows
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<h2 id="subset-columns-using-select">Subset columns using <code class="language-plaintext highlighter-rouge">select()</code></h2>

<p><a href="#contents"><em>Back to top</em></a></p>

<p>We use the <code class="language-plaintext highlighter-rouge">filter()</code> function to choose a subset of the rows from our data, but when we want to choose a subset of columns from our data we use <code class="language-plaintext highlighter-rouge">select()</code>. For example, if we only wanted to see the population (“pop”) and year values, we can do:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="n">pop</span><span class="p">,</span><span class="w"> </span><span class="n">year</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1,704 × 2
        pop  year
      &lt;dbl&gt; &lt;dbl&gt;
 1  8425333  1952
 2  9240934  1957
 3 10267083  1962
 4 11537966  1967
 5 13079460  1972
 6 14880372  1977
 7 12881816  1982
 8 13867957  1987
 9 16317921  1992
10 22227415  1997
# … with 1,694 more rows
</code></pre></div></div>

<p>We can also use <code class="language-plaintext highlighter-rouge">select()</code> to drop/remove particular columns by putting a minus sign (<code class="language-plaintext highlighter-rouge">-</code>) in front of the column name. For example, if we want everything but the continent column, we can do:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">continent</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1,704 × 5
   country      year      pop lifeExp gdpPercap
   &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;
 1 Afghanistan  1952  8425333    28.8      779.
 2 Afghanistan  1957  9240934    30.3      821.
 3 Afghanistan  1962 10267083    32.0      853.
 4 Afghanistan  1967 11537966    34.0      836.
 5 Afghanistan  1972 13079460    36.1      740.
 6 Afghanistan  1977 14880372    38.4      786.
 7 Afghanistan  1982 12881816    39.9      978.
 8 Afghanistan  1987 13867957    40.8      852.
 9 Afghanistan  1992 16317921    41.7      649.
10 Afghanistan  1997 22227415    41.8      635.
# … with 1,694 more rows
</code></pre></div></div>

<blockquote class="challenge">
  <h2 id="selecting-columns">selecting columns</h2>
  <p>Create a dataframe with only the <code class="language-plaintext highlighter-rouge">country</code>, <code class="language-plaintext highlighter-rouge">continent</code>, <code class="language-plaintext highlighter-rouge">year</code>, and <code class="language-plaintext highlighter-rouge">lifeExp</code> columns.</p>

  <blockquote class="solution">
    <h2 id="solution-7">Solution:</h2>
    <p>There are multiple ways to do this exercise. Here are two different possibilities.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="n">continent</span><span class="p">,</span><span class="w"> </span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1,704 × 4
   country     continent  year lifeExp
   &lt;chr&gt;       &lt;chr&gt;     &lt;dbl&gt;   &lt;dbl&gt;
 1 Afghanistan Asia       1952    28.8
 2 Afghanistan Asia       1957    30.3
 3 Afghanistan Asia       1962    32.0
 4 Afghanistan Asia       1967    34.0
 5 Afghanistan Asia       1972    36.1
 6 Afghanistan Asia       1977    38.4
 7 Afghanistan Asia       1982    39.9
 8 Afghanistan Asia       1987    40.8
 9 Afghanistan Asia       1992    41.7
10 Afghanistan Asia       1997    41.8
# … with 1,694 more rows
</code></pre></div>    </div>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">pop</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">gdpPercap</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1,704 × 4
   country      year continent lifeExp
   &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt;       &lt;dbl&gt;
 1 Afghanistan  1952 Asia         28.8
 2 Afghanistan  1957 Asia         30.3
 3 Afghanistan  1962 Asia         32.0
 4 Afghanistan  1967 Asia         34.0
 5 Afghanistan  1972 Asia         36.1
 6 Afghanistan  1977 Asia         38.4
 7 Afghanistan  1982 Asia         39.9
 8 Afghanistan  1987 Asia         40.8
 9 Afghanistan  1992 Asia         41.7
10 Afghanistan  1997 Asia         41.8
# … with 1,694 more rows
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<blockquote class="solution">
  <h2 id="bonus-using-helper-functions-with-select">Bonus: Using helper functions with <code class="language-plaintext highlighter-rouge">select()</code></h2>

  <p>The <code class="language-plaintext highlighter-rouge">select()</code> function has a bunch of helper functions that are handy if you are working with a dataset that has a lot of columns. You can see these helper functions on the <code class="language-plaintext highlighter-rouge">?select</code> help page. For example, let’s say we wanted to select the year column and all the columns that start with the letter “c”. You can do that with:</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">starts_with</span><span class="p">(</span><span class="s2">"c"</span><span class="p">))</span><span class="w">
</span></code></pre></div>  </div>

  <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1,704 × 3
    year country     continent
   &lt;dbl&gt; &lt;chr&gt;       &lt;chr&gt;    
 1  1952 Afghanistan Asia     
 2  1957 Afghanistan Asia     
 3  1962 Afghanistan Asia     
 4  1967 Afghanistan Asia     
 5  1972 Afghanistan Asia     
 6  1977 Afghanistan Asia     
 7  1982 Afghanistan Asia     
 8  1987 Afghanistan Asia     
 9  1992 Afghanistan Asia     
10  1997 Afghanistan Asia     
# … with 1,694 more rows
</code></pre></div>  </div>
  <p>This returns just the three columns we are interested in.</p>

  <blockquote class="challenge">
    <h2 id="using-select-with-a-helper-function">Using <code class="language-plaintext highlighter-rouge">select()</code> with a helper function</h2>

    <p>Find a helper function on the help page that will choose all the columns that have “p” as their last letter (ie: “pop”,”lifeExp”,”gdpPerCap”)</p>

    <blockquote class="solution">
      <h2 id="solution-8">Solution</h2>

      <p>The helper function <code class="language-plaintext highlighter-rouge">ends_with()</code> can help us here.</p>

      <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
</span><span class="n">select</span><span class="p">(</span><span class="n">ends_with</span><span class="p">(</span><span class="s2">"p"</span><span class="p">))</span><span class="w">
</span></code></pre></div>      </div>

      <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 1,704 × 3
        pop lifeExp gdpPercap
      &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;
 1  8425333    28.8      779.
 2  9240934    30.3      821.
 3 10267083    32.0      853.
 4 11537966    34.0      836.
 5 13079460    36.1      740.
 6 14880372    38.4      786.
 7 12881816    39.9      978.
 8 13867957    40.8      852.
 9 16317921    41.7      649.
10 22227415    41.8      635.
# … with 1,694 more rows
</code></pre></div>      </div>
    </blockquote>
  </blockquote>

</blockquote>

<h2 id="changing-the-shape-of-the-data">Changing the shape of the data</h2>

<p><a href="#contents"><em>Back to top</em></a></p>

<p>Data comes in many shapes and sizes, and one way we classify data is either “wide” or “long.” Data that is “long” has one row per observation. The gapminder_data data is in a long format. We have one row for each country for each year and each different measurement for that country is in a different column. We might describe this data as “tidy” because it makes it easy to work with <code class="language-plaintext highlighter-rouge">ggplot2</code> and <code class="language-plaintext highlighter-rouge">dplyr</code> functions (this is where the “tidy” in “tidyverse” comes from). As tidy as it may be, sometimes we may want our data in a “wide” format. Typically in “wide” format each row represents a group of observations and each value is placed in a different column rather than a different row. For example maybe we want only one row per country and want to spread the life expectancy values into different columns (one for each year).</p>

<p>The <code class="language-plaintext highlighter-rouge">tidyr</code> package contains the functions <code class="language-plaintext highlighter-rouge">pivot_wider</code> and <code class="language-plaintext highlighter-rouge">pivot_longer</code> that make it easy to switch between the two formats. The <code class="language-plaintext highlighter-rouge">tidyr</code> package is included in the <code class="language-plaintext highlighter-rouge">tidyverse</code> package so we don’t need to do anything to load it.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="n">continent</span><span class="p">,</span><span class="w"> </span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">lifeExp</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">pivot_wider</span><span class="p">(</span><span class="n">names_from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">values_from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lifeExp</span><span class="w"> </span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code># A tibble: 142 × 14
   country     continent `1952` `1957` `1962` `1967` `1972` `1977` `1982` `1987`
   &lt;chr&gt;       &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
 1 Afghanistan Asia        28.8   30.3   32.0   34.0   36.1   38.4   39.9   40.8
 2 Albania     Europe      55.2   59.3   64.8   66.2   67.7   68.9   70.4   72  
 3 Algeria     Africa      43.1   45.7   48.3   51.4   54.5   58.0   61.4   65.8
 4 Angola      Africa      30.0   32.0   34     36.0   37.9   39.5   39.9   39.9
 5 Argentina   Americas    62.5   64.4   65.1   65.6   67.1   68.5   69.9   70.8
 6 Australia   Oceania     69.1   70.3   70.9   71.1   71.9   73.5   74.7   76.3
 7 Austria     Europe      66.8   67.5   69.5   70.1   70.6   72.2   73.2   74.9
 8 Bahrain     Asia        50.9   53.8   56.9   59.9   63.3   65.6   69.1   70.8
 9 Bangladesh  Asia        37.5   39.3   41.2   43.5   45.3   46.9   50.0   52.8
10 Belgium     Europe      68     69.2   70.2   70.9   71.4   72.8   73.9   75.4
# … with 132 more rows, and 4 more variables: `1992` &lt;dbl&gt;, `1997` &lt;dbl&gt;,
#   `2002` &lt;dbl&gt;, `2007` &lt;dbl&gt;
</code></pre></div></div>

<p>Notice here that we tell <code class="language-plaintext highlighter-rouge">pivot_wider()</code> which columns to pull the names we wish our new columns to be named from the year variable, and the values to populate those columns from the lifeExp variable. (Again, neither of which have to be in quotes in the code when there are no special characters or spaces - certainly an incentive not to use special characters or spaces!) We see that the resulting table has new columns by year, and the values populate it with our remaining variables dictating the rows.</p>

<p>Before we move on to more data cleaning, let’s create the final gapminder dataframe we will be working with for the rest of the lesson!</p>

<blockquote class="challenge">
  <h2 id="final-americas-2007-gapminder-dataset">Final Americas 2007 gapminder dataset</h2>
  <p>Read in the <code class="language-plaintext highlighter-rouge">gapminder_data.csv</code> file, filter out the year 2007 and the continent “Americas.” Then drop the <code class="language-plaintext highlighter-rouge">year</code> and <code class="language-plaintext highlighter-rouge">continent</code> columns from the dataframe. Then save the new dataframe into a variable called <code class="language-plaintext highlighter-rouge">gapminder_data_2007</code>.</p>

  <blockquote class="solution">
    <h2 id="solution-9">Solution:</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data_2007</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/gapminder_data.csv"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">filter</span><span class="p">(</span><span class="n">year</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">2007</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">continent</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Americas"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">continent</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Rows: 1704 Columns: 6
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): country, continent
dbl (4): year, pop, lifeExp, gdpPercap

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<p>Awesome! This is the dataframe we will be using later on in this lesson.</p>

<h1 id="applying-it-to-your-own-data">Applying it to your own data</h1>

<p>Now it’s time to read in, clean, and make plots with your own data! 
Use your ideas from your brainstorming session yesterday to help you get started, but feel free to branch out and explore other things as well.
Let us know if you have questions; we’re here to help.</p>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>Package loading is an important first step in preparing an R environment.</p>
</li>
    
    <li><p>There are many useful functions in the <code class="language-plaintext highlighter-rouge">tidyverse</code> packages that can aid in data analysis.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="r-for-data-cleaning-and-merging" class="maintitle">R for Data Cleaning and Merging</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 75 min
      <br />
      <strong>Exercises:</strong> 15 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How can I combine two datasets from different sources?</p>
</li>
	
	<li><p>How can data tidying facilitate answering analysis questions?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>To become familiar with more functions of the <code class="language-plaintext highlighter-rouge">dplyr</code> and <code class="language-plaintext highlighter-rouge">tidyr</code> packages.</p>
</li>
	
	<li><p>To be able to use <code class="language-plaintext highlighter-rouge">dplyr</code> and <code class="language-plaintext highlighter-rouge">tidyr</code> to prepare data for analysis.</p>
</li>
	
	<li><p>To be able to combine two different data sources using joins.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<h3 id="contents">Contents</h3>

<ol>
  <li><a href="#cleaning-up-data">Cleaning up data</a></li>
  <li><a href="#joining-data-frames">Joining data frames</a></li>
  <li><a href="#analyzing-combined-data">Analyzing combined data</a></li>
</ol>

<h1 id="cleaning-up-data">Cleaning up data</h1>

<p><a href="#contents"><em>Back to top</em></a></p>

<p>Researchers are often pulling data from several sources, and the process of making data compatible with one another and prepared for analysis can be a large undertaking. Luckily, there are many functions that allow us to do this in R. We’ve been working with the gapminder dataset, which contains population and GDP data by year. In this section, we practice cleaning and preparing a second dataset containing CO2 emissions data by country and year, sourced from <a href="https://data.un.org/_Docs/SYB/CSV/SYB63_310_202009_Carbon%20Dioxide%20Emission%20Estimates.csv">the UN</a>.</p>

<p>It’s always good to go into data cleaning with a clear goal in mind. Here, we’d like to prepare the CO2 UN data to be compatible with our gapminder data so we can directly compare GDP to CO2 emissions. To make this work, we’d like a data frame that contains a column with the country name, and columns for different ways of measuring CO2 emissions. We will also want the data to be collected as close to 2007 as possible (the last year we have data for in gapminder). Let’s start with reading the data in using <code class="language-plaintext highlighter-rouge">read_csv()</code></p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/co2-un-data.csv"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in read_csv("data/co2-un-data.csv"): could not find function "read_csv"
</code></pre></div></div>

<p>The output gives us a warning about missing column names being filled in with things like ‘X3’, ‘X4’, etc. Looking at the table that is outputted by <code class="language-plaintext highlighter-rouge">read_csv()</code> we can see that there appear to be two rows at the top of the file that contain information about the data in the table. The first is a header that tells us the table number and its name. Ideally, we’d skip that. We can do this using the <code class="language-plaintext highlighter-rouge">skip=</code> argument in read_csv by giving it a number of lines to skip.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/co2-un-data.csv"</span><span class="p">,</span><span class="w"> </span><span class="n">skip</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in read_csv("data/co2-un-data.csv", skip = 1): could not find function "read_csv"
</code></pre></div></div>

<p>Now we get a similar Warning message as before, but the outputted table looks better.</p>

<blockquote>
  <h2 id="warnings-and-errors">Warnings and Errors</h2>
  <p>It’s important to differentiate between Warnings and Errors in R. A warning tells us, “you might want to know about this issue, but R still did what you asked”. An error tells us, “there’s something wrong with your code or your data and R didn’t do what you asked”. You need to fix any errors that arise. Warnings, are probably best to resolve or at least understand why they are coming up.
{.callout}</p>
</blockquote>

<p>We can resolve this warning by telling <code class="language-plaintext highlighter-rouge">read_csv()</code> what the column names should be with the <code class="language-plaintext highlighter-rouge">col_names()</code> argument where we give it the column names we want within the c() function separated by commas. If we do this, then we need to set skip to 2 to also skip the column headings. Let’s also save this dataframe to <code class="language-plaintext highlighter-rouge">co2_emissions_dirty</code> so that we don’t have to read it in every time we want to clean it even more.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">co2_emissions_dirty</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/co2-un-data.csv"</span><span class="p">,</span><span class="w"> </span><span class="n">skip</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w">
         </span><span class="n">col_names</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"region"</span><span class="p">,</span><span class="w"> </span><span class="s2">"country"</span><span class="p">,</span><span class="w"> </span><span class="s2">"year"</span><span class="p">,</span><span class="w"> </span><span class="s2">"series"</span><span class="p">,</span><span class="w"> </span><span class="s2">"value"</span><span class="p">,</span><span class="w"> </span><span class="s2">"footnotes"</span><span class="p">,</span><span class="w"> </span><span class="s2">"source"</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in read_csv("data/co2-un-data.csv", skip = 2, col_names = c("region", : could not find function "read_csv"
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">co2_emissions_dirty</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in eval(expr, envir, enclos): object 'co2_emissions_dirty' not found
</code></pre></div></div>

<blockquote class="solution">
  <h2 id="bonus-another-way-to-deal-with-this-error">Bonus: Another way to deal with this error</h2>

  <p>There are often multiple ways to clean data. Here we  read in the table, get the warning and then fix the column names using the rename function.</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/co2-un-data.csv"</span><span class="p">,</span><span class="w"> </span><span class="n">skip</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">rename</span><span class="p">(</span><span class="n">country</span><span class="o">=</span><span class="n">X2</span><span class="p">)</span><span class="w">
</span></code></pre></div>  </div>

  <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in read_csv("data/co2-un-data.csv", skip = 1) %&gt;% rename(country = X2): could not find function "%&gt;%"
</code></pre></div>  </div>

  <p>Many data analysts prefer to have their column headings and variable names be in all lower case. We can use a variation of <code class="language-plaintext highlighter-rouge">rename()</code>, which is <code class="language-plaintext highlighter-rouge">rename_all()</code> that allows us to set all of the column headings to lower case by giving it the name of the tolower function, which makes everything lowercase.</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/co2-un-data.csv"</span><span class="p">,</span><span class="w"> </span><span class="n">skip</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
 </span><span class="n">rename_all</span><span class="p">(</span><span class="n">tolower</span><span class="p">)</span><span class="w">
</span></code></pre></div>  </div>

  <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in read_csv("data/co2-un-data.csv", skip = 1) %&gt;% rename_all(tolower): could not find function "%&gt;%"
</code></pre></div>  </div>
</blockquote>

<p>We previously saw how we can subset columns from a data frame using the select function. There are a lot of columns with extraneous information in this dataset, let’s subset out the columns we are interested in.</p>

<blockquote class="challenge">
  <h2 id="reviewing-selecting-columns">Reviewing selecting columns</h2>
  <p>Select the country, year, series, and value columns from our dataset.</p>

  <blockquote class="solution">
    <h2 id="solution">Solution:</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">co2_emissions_dirty</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in co2_emissions_dirty %&gt;% select(country, year, series, value): could not find function "%&gt;%"
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<p>The series column has two methods of quantifying CO2 emissions - “Emissions (thousand metric tons of carbon dioxide)” and “Emissions per capita (metric tons of carbon dioxide)”. Those are long titles that we’d like to shorten to make them easier to work with. We can shorten them to “total_emissions” and “per_capita_emissions” using the recode function. We need to do this within the mutate function where we will mutate the series column. The syntax in the recode function is to tell recode which column we want to recode and then what the old value (e.g. “Emissions (thousand metric tons of carbon dioxide)”) should equal after recoding (e.g. “total”).</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">co2_emissions_dirty</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">select</span><span class="p">(</span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">series</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">recode</span><span class="p">(</span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="s2">"Emissions (thousand metric tons of carbon dioxide)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"total_emissions"</span><span class="p">,</span><span class="w">
                         </span><span class="s2">"Emissions per capita (metric tons of carbon dioxide)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"per_capita_emissions"</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in co2_emissions_dirty %&gt;% select(country, year, series, value) %&gt;% : could not find function "%&gt;%"
</code></pre></div></div>

<p>Recall that we’d like to have separate columns for the two ways that we CO2 emissions data. To achieve this, we’ll use the pivot_wider function that we saw previously. The columns we want to spread out are series (i.e. names_from) and value (i.e. values_from).</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">co2_emissions_dirty</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">series</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">recode</span><span class="p">(</span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="s2">"Emissions (thousand metric tons of carbon dioxide)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"total_emission"</span><span class="p">,</span><span class="w">
                         </span><span class="s2">"Emissions per capita (metric tons of carbon dioxide)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"per_capita_emission"</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">pivot_wider</span><span class="p">(</span><span class="n">names_from</span><span class="o">=</span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="n">values_from</span><span class="o">=</span><span class="n">value</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in co2_emissions_dirty %&gt;% select(country, year, series, value) %&gt;% : could not find function "%&gt;%"
</code></pre></div></div>

<p>Excellent! The last step before we can join this data frame is to get the most data that is for the year closest to 2007 so we can make a more direct comparison to the most recent data we have from gapminder. For the sake of time, we’ll just tell you that we want data from 2005.</p>

<blockquote class="solution">
  <h2 id="bonus-how-did-we-determine-that-2005-is-the-closest-year-to-2007">Bonus: How did we determine that 2005 is the closest year to 2007?</h2>

  <p>We want to make sure we pick a year that is close to 2005, but also a year that has a decent amount of data to work with. One useful tool is the <code class="language-plaintext highlighter-rouge">count()</code> function, which will tell us how many times a value is repeated in a column of a data frame. Let’s use this function on the year column to see which years we have data for and to tell us whether we have a good number of countries represented in that year.</p>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">co2_emissions_dirty</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
 </span><span class="n">select</span><span class="p">(</span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
 </span><span class="n">mutate</span><span class="p">(</span><span class="n">series</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">recode</span><span class="p">(</span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="s2">"Emissions (thousand metric tons of carbon dioxide)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"total"</span><span class="p">,</span><span class="w">
                        </span><span class="s2">"Emissions per capita (metric tons of carbon dioxide)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"per_capita"</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">pivot_wider</span><span class="p">(</span><span class="n">names_from</span><span class="o">=</span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="n">values_from</span><span class="o">=</span><span class="n">value</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
 </span><span class="n">count</span><span class="p">(</span><span class="n">year</span><span class="p">)</span><span class="w">
</span></code></pre></div>  </div>

  <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in co2_emissions_dirty %&gt;% select(country, year, series, value) %&gt;% : could not find function "%&gt;%"
</code></pre></div>  </div>

  <p>It looks like we have data for 140 countries in 2005 and 2010. We chose 2005 because it is closer to 2007.</p>
</blockquote>

<blockquote class="challenge">
  <h2 id="filtering-rows-and-removing-columns">Filtering rows and removing columns</h2>
  <p>Filter out data from 2005 and then drop the year column. (Since we will have only data from one year, it is now irrelevant.)</p>

  <blockquote class="solution">
    <h2 id="solution-1">Solution:</h2>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">co2_emissions_dirty</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
 </span><span class="n">select</span><span class="p">(</span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
 </span><span class="n">mutate</span><span class="p">(</span><span class="n">series</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">recode</span><span class="p">(</span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="s2">"Emissions (thousand metric tons of carbon dioxide)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"total"</span><span class="p">,</span><span class="w">
                        </span><span class="s2">"Emissions per capita (metric tons of carbon dioxide)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"per_capita"</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
 </span><span class="n">pivot_wider</span><span class="p">(</span><span class="n">names_from</span><span class="o">=</span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="n">values_from</span><span class="o">=</span><span class="n">value</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
 </span><span class="n">filter</span><span class="p">(</span><span class="n">year</span><span class="o">==</span><span class="m">2005</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
 </span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">year</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in co2_emissions_dirty %&gt;% select(country, year, series, value) %&gt;% : could not find function "%&gt;%"
</code></pre></div>    </div>
  </blockquote>
</blockquote>

<p>Finally, let’s go ahead and assign the output of this code chunk, which is the cleaned dataframe, to a variable name:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">co2_emissions</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">co2_emissions_dirty</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">series</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">recode</span><span class="p">(</span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="s2">"Emissions (thousand metric tons of carbon dioxide)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"total_emission"</span><span class="p">,</span><span class="w">
                         </span><span class="s2">"Emissions per capita (metric tons of carbon dioxide)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"per_capita_emission"</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">pivot_wider</span><span class="p">(</span><span class="n">names_from</span><span class="o">=</span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="n">values_from</span><span class="o">=</span><span class="n">value</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">filter</span><span class="p">(</span><span class="n">year</span><span class="o">==</span><span class="m">2005</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">year</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in co2_emissions_dirty %&gt;% select(country, year, series, value) %&gt;% : could not find function "%&gt;%"
</code></pre></div></div>

<blockquote>
  <p><strong>Looking at your data:</strong> You can get a look at your data-cleaning hard work by navigating to the <strong>Environment</strong> tab in RStudio and clicking the table icon next to the variable name. Notice when we do this, RStudio automatically runs the <code class="language-plaintext highlighter-rouge">View()</code> command. We’ve made a lot of progress!
{.callout}</p>
</blockquote>

<h1 id="joining-data-frames">Joining data frames</h1>

<p><a href="#contents"><em>Back to top</em></a></p>

<p>Now we’re ready to join our CO2 emissions data to the gapminder data. Previously we saw that we could read in and filter the gapminder data like this to get the data from the Americas for 2007 so we can create a new dataframe with our filtered data:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data_2007</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/gapminder_data.csv"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">filter</span><span class="p">(</span><span class="n">year</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">2007</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">continent</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Americas"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">continent</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in read_csv("data/gapminder_data.csv") %&gt;% filter(year == 2007 &amp; : could not find function "%&gt;%"
</code></pre></div></div>

<p>Look at the data in <code class="language-plaintext highlighter-rouge">co2_emissions</code> and <code class="language-plaintext highlighter-rouge">gapminder_data_2007</code>. If you had to merge these two data frames together, which column would you use to merge them together? If you said “country” - good job!</p>

<p>We’ll call country our “key”. Now, when we join them together, can you think of any problems we might run into when we merge things? We might not have CO2 emissions data for all of the countries in the gapminder dataset and vice versa. Also, a country might be represented in both data frames but not by the same name in both places. As an example, write down the name of the country that the University of Michigan is in - we’ll come back to your answer shortly!</p>

<p>The dplyr package has a number of tools for joining data frames together depending on what we want to do with the rows of the data of countries that are not represented in both data frames. Here we’ll be using <code class="language-plaintext highlighter-rouge">inner_join()</code> and <code class="language-plaintext highlighter-rouge">anti_join()</code>.</p>

<p>In an “inner join”, the new data frame only has those rows where the same key is found in both data frames. This is a very commonly used join.</p>

<p><img src="../fig/r-data-analysis/join-inner.png" alt="" /></p>

<blockquote class="solution">
  <h2 id="bonus-other-dplyr-join-functions">Bonus: Other dplyr join functions</h2>

  <p>Outer joins and can be performed using <code class="language-plaintext highlighter-rouge">left_join()</code>, <code class="language-plaintext highlighter-rouge">right_join()</code>, and <code class="language-plaintext highlighter-rouge">full_join()</code>. In a “left join”, if the key is present in the left hand data frame, it will appear in the output, even if it is not found in the the right hand data frame. For a right join, the opposite is true. For a full join, all possible keys are included in the output data frame.</p>

  <p><img src="../fig/r-data-analysis/join-outer.png" alt="" /></p>
</blockquote>

<p>If you don’t already have it read in, read in the gapminder data:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s1">'data/gapminder_data.csv'</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in read_csv("data/gapminder_data.csv"): could not find function "read_csv"
</code></pre></div></div>

<p>Let’s give the <code class="language-plaintext highlighter-rouge">inner_join()</code> function a try.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">inner_join</span><span class="p">(</span><span class="n">gapminder_data</span><span class="p">,</span><span class="w"> </span><span class="n">co2_emissions</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in inner_join(gapminder_data, co2_emissions): could not find function "inner_join"
</code></pre></div></div>

<p>Do you see that we now have data from both data frames joined together in the same data frame? One thing to note about the output is that <code class="language-plaintext highlighter-rouge">inner_join()</code> tells us that that it joined by “country”. We can make this explicit using the “by” argument in the join functions</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">inner_join</span><span class="p">(</span><span class="n">gapminder_data</span><span class="p">,</span><span class="w"> </span><span class="n">co2_emissions</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s2">"country"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in inner_join(gapminder_data, co2_emissions, by = "country"): could not find function "inner_join"
</code></pre></div></div>

<p>One thing to notice is that gapminder data had 25 rows, but the output of our join only had 21. Let’s investigate. It appears that there must have been countries in the gapminder data that did not appear in our co2_emissions data frame.</p>

<p>Let’s use <code class="language-plaintext highlighter-rouge">anti_join()</code> for this - this will show us the data for the keys on the left that are missing from the data frame on the right.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">anti_join</span><span class="p">(</span><span class="n">gapminder_data</span><span class="p">,</span><span class="w"> </span><span class="n">co2_emissions</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s2">"country"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in anti_join(gapminder_data, co2_emissions, by = "country"): could not find function "anti_join"
</code></pre></div></div>
<p>We can see that the co2_emissions data were missing for Bolivia, Puerto Rico, United States, and Venezuela.</p>

<p>If we look at the co2_emissions data with <code class="language-plaintext highlighter-rouge">View()</code>, we will see that Bolivia, United States, and Venezuela are called different things in the co2_emissions data frame. They’re called “Bolivia (Plurin. State of)”, “United States of America”, and “Venezuela (Boliv. Rep. of)”. Puerto Rico isn’t a country; it’s part of the United States. Using <code class="language-plaintext highlighter-rouge">mutate()</code> and <code class="language-plaintext highlighter-rouge">recode()</code>, we can re-import the co2_emissions data so that the country names for Bolivia, United States, and Venezuela, match those in the gapminder data.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">co2_emissions</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/co2-un-data.csv"</span><span class="p">,</span><span class="w"> </span><span class="n">skip</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w">
                          </span><span class="n">col_names</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"region"</span><span class="p">,</span><span class="w"> </span><span class="s2">"country"</span><span class="p">,</span><span class="w"> </span><span class="s2">"year"</span><span class="p">,</span><span class="w">
                                      </span><span class="s2">"series"</span><span class="p">,</span><span class="w"> </span><span class="s2">"value"</span><span class="p">,</span><span class="w"> </span><span class="s2">"footnotes"</span><span class="p">,</span><span class="w"> </span><span class="s2">"source"</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">series</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">recode</span><span class="p">(</span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="s2">"Emissions (thousand metric tons of carbon dioxide)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"total"</span><span class="p">,</span><span class="w">
                         </span><span class="s2">"Emissions per capita (metric tons of carbon dioxide)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"per_capita"</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">pivot_wider</span><span class="p">(</span><span class="n">names_from</span><span class="o">=</span><span class="n">series</span><span class="p">,</span><span class="w"> </span><span class="n">values_from</span><span class="o">=</span><span class="n">value</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">filter</span><span class="p">(</span><span class="n">year</span><span class="o">==</span><span class="m">2005</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">year</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">country</span><span class="o">=</span><span class="n">recode</span><span class="p">(</span><span class="n">country</span><span class="p">,</span><span class="w">
                        </span><span class="s2">"Bolivia (Plurin. State of)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Bolivia"</span><span class="p">,</span><span class="w">
                        </span><span class="s2">"United States of America"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"United States"</span><span class="p">,</span><span class="w">
                        </span><span class="s2">"Venezuela (Boliv. Rep. of)"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Venezuela"</span><span class="p">)</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in read_csv("data/co2-un-data.csv", skip = 2, col_names = c("region", : could not find function "%&gt;%"
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">anti_join</span><span class="p">(</span><span class="n">gapminder_data</span><span class="p">,</span><span class="w"> </span><span class="n">co2_emissions</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s2">"country"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in anti_join(gapminder_data, co2_emissions, by = "country"): could not find function "anti_join"
</code></pre></div></div>

<p>Now we see that our recode enabled the join for all countries in the gapminder, and we are left with Puerto Rico. In the next exercise, let’s recode Puerto Rico as United States in the gapminder data and then use <code class="language-plaintext highlighter-rouge">group_by()</code> and <code class="language-plaintext highlighter-rouge">summarize()</code> to aggregate the data; we’ll use the population data to weight the life expectancy and GDP values.</p>

<p>In the gapminder data, let’s recode Puerto Rico as United States.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/gapminder_data.csv"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
</span><span class="n">filter</span><span class="p">(</span><span class="n">year</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">2007</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">continent</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Americas"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
</span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
</span><span class="n">mutate</span><span class="p">(</span><span class="n">country</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">recode</span><span class="p">(</span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="s2">"Puerto Rico"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"United States"</span><span class="p">))</span><span class="w"> 
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in read_csv("data/gapminder_data.csv") %&gt;% filter(year == 2007 &amp; : could not find function "%&gt;%"
</code></pre></div></div>

<p>Now we have to group Puerto Rico and the US together, aggregating and calculating the data for all of the other columns. This is a little tricky - we will need a weighted average of lifeExp and gdpPercap. You can do this using standard order of operations.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/gapminder_data.csv"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">filter</span><span class="p">(</span><span class="n">year</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">2007</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">continent</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Americas"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">year</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">continent</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">country</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">recode</span><span class="p">(</span><span class="n">country</span><span class="p">,</span><span class="w"> </span><span class="s2">"Puerto Rico"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"United States"</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">group_by</span><span class="p">(</span><span class="n">country</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarize</span><span class="p">(</span><span class="n">lifeExp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">lifeExp</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">pop</span><span class="p">)</span><span class="o">/</span><span class="nf">sum</span><span class="p">(</span><span class="n">pop</span><span class="p">),</span><span class="w">
            </span><span class="n">gdpPercap</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">gdpPercap</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">pop</span><span class="p">)</span><span class="o">/</span><span class="nf">sum</span><span class="p">(</span><span class="n">pop</span><span class="p">),</span><span class="w">
            </span><span class="n">pop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in read_csv("data/gapminder_data.csv") %&gt;% filter(year == 2007 &amp; : could not find function "%&gt;%"
</code></pre></div></div>

<p>Let’s check to see if it worked!</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">anti_join</span><span class="p">(</span><span class="n">gapminder_data</span><span class="p">,</span><span class="w"> </span><span class="n">co2_emissions</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s2">"country"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in anti_join(gapminder_data, co2_emissions, by = "country"): could not find function "anti_join"
</code></pre></div></div>
<p>Now our <code class="language-plaintext highlighter-rouge">anti_join()</code> returns an empty data frame, which tells us that we have reconciled all of the keys from the gapminder data with the data in the co2_emissions data frame.</p>

<p>Finally, let’s use the <code class="language-plaintext highlighter-rouge">inner_join()</code> to create a new data frame:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_co2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">inner_join</span><span class="p">(</span><span class="n">gapminder_data</span><span class="p">,</span><span class="w"> </span><span class="n">co2_emissions</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s2">"country"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in inner_join(gapminder_data, co2_emissions, by = "country"): could not find function "inner_join"
</code></pre></div></div>

<p>One last thing! What if we’re interested in distinguishing between countries in North America and South America? We want to create two groups - Canada, the United States, and Mexico in one and the other countries in another.</p>

<p>We can create a grouping variable using <code class="language-plaintext highlighter-rouge">mutate()</code> combined with an <code class="language-plaintext highlighter-rouge">if_else()</code> function - a very useful pairing.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_co2</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">  
</span><span class="n">mutate</span><span class="p">(</span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">if_else</span><span class="p">(</span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Canada"</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"United States"</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Mexico"</span><span class="p">,</span><span class="w"> </span><span class="s2">"north"</span><span class="p">,</span><span class="w"> </span><span class="s2">"south"</span><span class="p">))</span><span class="w">  
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in gapminder_co2 %&gt;% mutate(region = if_else(country == "Canada" | : could not find function "%&gt;%"
</code></pre></div></div>
<p>Let’s look at the output - see how the Canada, US, and Mexico rows are all labeled as “north” and everything else is labeled as “south”</p>

<p>We have reached our data cleaning goals! One of the best aspects of doing all of these steps coded in R is that our efforts are reproducible, and the raw data is maintained. With good documentation of data cleaning and analysis steps, we could easily share our work with another researcher who would be able to repeat what we’ve done. However, it’s also nice to have a saved <code class="language-plaintext highlighter-rouge">csv</code> copy of our clean data. That way we can access it later without needing to redo our data cleaning, and we can also share the cleaned data with collaborators. To save our dataframe, we’ll use <code class="language-plaintext highlighter-rouge">write_csv()</code>.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">write_csv</span><span class="p">(</span><span class="n">gapminder_co2</span><span class="p">,</span><span class="w"> </span><span class="s2">"data/gapminder_co2.csv"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in write_csv(gapminder_co2, "data/gapminder_co2.csv"): could not find function "write_csv"
</code></pre></div></div>

<p>Great - Now we can move on to the analysis!</p>

<h1 id="analyzing-combined-data">Analyzing combined data</h1>

<p><a href="#contents"><em>Back to top</em></a></p>

<p>For our analysis, we have two questions we’d like to answer. First, is there a relationship between the GDP of a country and the amount of CO2 emitted (per capita)? Second, Canada, the United States, and Mexico account for nearly half of the population of the Americas. What percent of the total CO2 production do they account for?</p>

<p>To answer the first question, we’ll plot the CO2 emitted (on a per capita basis) against the GDP (on a per capita basis) using a scatter plot:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder_co2</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">gdpPercap</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="n">per_capita</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">"GDP (per capita)"</span><span class="p">,</span><span class="w">
       </span><span class="n">y</span><span class="o">=</span><span class="s2">"CO2 emitted (per capita)"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="o">=</span><span class="s2">"There is a strong association between a nation's GDP \nand the amount of CO2 it produces"</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in ggplot(gapminder_co2, aes(x = gdpPercap, y = per_capita)): could not find function "ggplot"
</code></pre></div></div>

<p><em>Tip:</em> Notice we used the <code class="language-plaintext highlighter-rouge">\n</code> in our title to get a new line to prevent it from getting cut off.</p>

<p>To help clarify the association, we can add a fit line through the data using <code class="language-plaintext highlighter-rouge">geom_smooth()</code></p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder_co2</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">gdpPercap</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="n">per_capita</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">"GDP (per capita)"</span><span class="p">,</span><span class="w">
       </span><span class="n">y</span><span class="o">=</span><span class="s2">"CO2 emitted (per capita)"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="o">=</span><span class="s2">"There is a strong association between a nation's GDP \nand the amount of CO2 it produces"</span><span class="w">
  </span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in ggplot(gapminder_co2, aes(x = gdpPercap, y = per_capita)): could not find function "ggplot"
</code></pre></div></div>

<p>We can force the line to be straight using <code class="language-plaintext highlighter-rouge">method="lm"</code> as an argument to <code class="language-plaintext highlighter-rouge">geom_smooth</code></p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">gapminder_co2</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">gdpPercap</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="n">per_capita</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">"GDP (per capita)"</span><span class="p">,</span><span class="w">
       </span><span class="n">y</span><span class="o">=</span><span class="s2">"CO2 emitted (per capita)"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="o">=</span><span class="s2">"There is a strong association between a nation's GDP \nand the amount of CO2 it produces"</span><span class="w">
  </span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s2">"lm"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in ggplot(gapminder_co2, aes(x = gdpPercap, y = per_capita)): could not find function "ggplot"
</code></pre></div></div>

<p>To answer our first question, as the title of our plot indicates there is indeed a strong association between a nation’s GDP and the amount of CO2 it produces.</p>

<p>For the second question, we want to create two groups - Canada, the United States, and Mexico in one and the other countries in another.</p>

<p>We can create a grouping variable using <code class="language-plaintext highlighter-rouge">mutate()</code> combined with an <code class="language-plaintext highlighter-rouge">if_else()</code> function - a very useful pairing.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_co2</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">if_else</span><span class="p">(</span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Canada"</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"United States"</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Mexico"</span><span class="p">,</span><span class="w"> </span><span class="s2">"north"</span><span class="p">,</span><span class="w"> </span><span class="s2">"south"</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in gapminder_co2 %&gt;% mutate(region = if_else(country == "Canada" | : could not find function "%&gt;%"
</code></pre></div></div>

<p>Now we can use this column to repeat our <code class="language-plaintext highlighter-rouge">group_by()</code> and <code class="language-plaintext highlighter-rouge">summarize()</code> steps</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_co2</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">if_else</span><span class="p">(</span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Canada"</span><span class="w"> </span><span class="o">|</span><span class="w">
                            </span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"United States"</span><span class="w"> </span><span class="o">|</span><span class="w">
                            </span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Mexico"</span><span class="p">,</span><span class="w"> </span><span class="s2">"north"</span><span class="p">,</span><span class="w"> </span><span class="s2">"south"</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">group_by</span><span class="p">(</span><span class="n">region</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarize</span><span class="p">(</span><span class="n">sumtotal</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">total</span><span class="p">),</span><span class="w">
            </span><span class="n">sumpop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">pop</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in gapminder_co2 %&gt;% mutate(region = if_else(country == "Canada" | : could not find function "%&gt;%"
</code></pre></div></div>

<p>The <code class="language-plaintext highlighter-rouge">if_else()</code> statement reads like, “if country equals “Canada” OR <code class="language-plaintext highlighter-rouge">|</code> “United states” OR “Mexico”, the new variable region should be “north”, else “south””. It’s worth exploring logical operators for “or” <code class="language-plaintext highlighter-rouge">|</code>, “and” <code class="language-plaintext highlighter-rouge">&amp;&amp;</code>, and “not” <code class="language-plaintext highlighter-rouge">!</code>, which opens up a great deal of possibilities for writing code to do what you want.</p>

<p>We see that although Canada, the United States, and Mexico account for close to half the population of the Americas, they account for 88% of the CO2 emitted. We just did this math quickly by plugging the numbers from our table into the console to get the percentages. Can we make that a little more reproducible by calculating percentages for population (pop) and total emissions (total) into our data before summarizing?</p>

<h1 id="bonus">Bonus</h1>

<h2 id="bonus-exercise">Bonus exercise</h2>

<blockquote class="challenge">
  <h2 id="calculating-percent">Calculating percent</h2>

  <p>What percentage of the population and CO2 emissions in the Americas does the United States make up? What percentage of the population and CO2 emission does North America make up?</p>

  <blockquote class="solution">
    <h2 id="solution-2">Solution</h2>

    <p>Create a new variable using <code class="language-plaintext highlighter-rouge">mutate()</code> that calculates percentages for the pop and total variables.</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_co2</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">if_else</span><span class="p">(</span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Canada"</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"United States"</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Mexico"</span><span class="p">,</span><span class="w"> </span><span class="s2">"north"</span><span class="p">,</span><span class="w"> </span><span class="s2">"south"</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">totalPercent</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">total</span><span class="o">/</span><span class="nf">sum</span><span class="p">(</span><span class="n">total</span><span class="p">)</span><span class="o">*</span><span class="m">100</span><span class="p">,</span><span class="w">
         </span><span class="n">popPercent</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="o">/</span><span class="nf">sum</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span><span class="o">*</span><span class="m">100</span><span class="p">)</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in gapminder_co2 %&gt;% mutate(region = if_else(country == "Canada" | : could not find function "%&gt;%"
</code></pre></div>    </div>

    <p>This table shows that the United states makes up 33% of the population of the Americas, but accounts for 77% of total emissions. Now let’s take a look at population and emission for the two different continents:</p>

    <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_co2</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">region</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">if_else</span><span class="p">(</span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Canada"</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"United States"</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">country</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Mexico"</span><span class="p">,</span><span class="w"> </span><span class="s2">"north"</span><span class="p">,</span><span class="w"> </span><span class="s2">"south"</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">totalPercent</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">total</span><span class="o">/</span><span class="nf">sum</span><span class="p">(</span><span class="n">total</span><span class="p">)</span><span class="o">*</span><span class="m">100</span><span class="p">,</span><span class="w">
         </span><span class="n">popPercent</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pop</span><span class="o">/</span><span class="nf">sum</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span><span class="o">*</span><span class="m">100</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">group_by</span><span class="p">(</span><span class="n">region</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarize</span><span class="p">(</span><span class="n">sumTotalPercent</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">totalPercent</span><span class="p">),</span><span class="w">
            </span><span class="n">sumPopPercent</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">popPercent</span><span class="p">))</span><span class="w">
</span></code></pre></div>    </div>

    <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in gapminder_co2 %&gt;% mutate(region = if_else(country == "Canada" | : could not find function "%&gt;%"
</code></pre></div>    </div>

  </blockquote>
</blockquote>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>Assessing data source and structure is an important first step in analysis.</p>
</li>
    
    <li><p>Data analysis in R facilitates reproducible research.</p>
</li>
    
    <li><p>Preparing data for analysis can take significant effort and planning.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="writing-reports-with-r-markdown" class="maintitle">Writing Reports with R Markdown</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 105 min
      <br />
      <strong>Exercises:</strong> 30 min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>How can I make reproducible reports using R Markdown?</p>
</li>
	
	<li><p>How do I format text using Markdown?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>To create a report in R Markdown that combines text, code, and figures.</p>
</li>
	
	<li><p>To use Markdown to format our report.</p>
</li>
	
	<li><p>To understand how to use R code chunks to include or hide code, figures, and messages.</p>
</li>
	
	<li><p>To be aware of the various report formats that can be rendered using R Markdown.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<h3 id="contents">Contents</h3>
<ol>
  <li><a href="#r-for-data-analysis-review">R for data analysis review</a></li>
  <li><a href="#why-use-r-markdown">What is R Markdown and why use it?</a></li>
  <li><a href="#creating-a-reports-directory">Creating a reports directory</a></li>
  <li><a href="#creating-an-r-markdown-file">Creating an R Markdown file</a></li>
  <li><a href="#basic-components-of-r-markdown">Basic components of R Markdown</a>
    <ul>
      <li><a href="#header">Header</a></li>
      <li><a href="#code-chunks">Code chunks</a></li>
      <li><a href="#text">Text</a></li>
    </ul>
  </li>
  <li><a href="#starting-the-report">Starting the report</a></li>
  <li><a href="#formatting">Formatting</a></li>
  <li><a href="#applying-it-to-your-own-data">Applying it to your own data</a></li>
</ol>

<h2 id="r-for-data-analysis-review">R for data analysis review</h2>

<p>Remember that yesterday we made a scatter plot of year vs. population, separated into a plot for each contient, 
and that it had 2 outliers? Which countries are those?</p>

<blockquote class="solution">
  <h2 id="solution">Solution</h2>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="n">filter</span><span class="p">(</span><span class="n">pop</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">1e9</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="n">select</span><span class="p">(</span><span class="n">country</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="n">unique</span><span class="p">()</span><span class="w">
</span></code></pre></div>  </div>

  <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in filter(., pop &gt; 1e+09): object 'gapminder' not found
</code></pre></div>  </div>
</blockquote>

<p>Next, plot year vs. population separated into a plot for each continent but excluding the 2 outlier countries.</p>

<blockquote class="solution">
  <h2 id="solution-1">Solution</h2>

  <div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="n">filter</span><span class="p">(</span><span class="n">country</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s1">'China'</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">country</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s1">'India'</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">year</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="n">pop</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
</span><span class="n">geom_point</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
</span><span class="n">facet_wrap</span><span class="p">(</span><span class="n">vars</span><span class="p">(</span><span class="n">continent</span><span class="p">))</span><span class="w">
</span></code></pre></div>  </div>

  <div class="language-plaintext error highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Error in filter(., country != "China" &amp; country != "India"): object 'gapminder' not found
</code></pre></div>  </div>
</blockquote>

<h2 id="what-is-r-markdown-and-why-use-it-why-use-r-markdown">What is R Markdown and why use it? {##why-use-r-markdown}</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>Recall that our  goal is to generate a report to the United Nations on how a country’s life expectancy is related to GDP.</p>

<blockquote class="discussion">
  <h2 id="discusion">Discusion</h2>
  <p>How do you usually share data analyses with your collaborators? Many people share them through a Word or PDF document, a spreadsheet, slides, a  graphic, etc.</p>
</blockquote>

<p>In R Markdown, you can incorporate ordinary text (ex. experimental methods, analysis and discussion of results) alongside code and figures! (Some people write entire manuscripts in R Markdown.) This is useful for writing reproducible reports and publications, sharing work with collaborators, writing up homework, and keeping a bioinformatics notebook. Because the code is emedded in the document, the tables and figures are <em>reproducible</em>. Anyone can run the code and get the same results. If you find an error or want to add more to the report, you can just re-run the document and you’ll have updated tables and figures! This concept of combining text and code is called “literate programming”. To do this we use R Markdown, which combines Markdown (renders plain text) with R. You can output an html, PDF, or Word document that you can share with others. In fact, this webpage is an example of a rendered R markdown file!</p>

<p>(If you are familiar with Jupyter notebooks in the Python programming environment, R Markdown is  R’s equivalent of a Jupyter notebook.)</p>

<h2 id="creating-an-r-markdown-file">Creating an R Markdown file</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>Now that we have a better understanding of what we can use R Markdown files for, let’s start writing a report!</p>

<p>To create an R Markdown file:</p>
<ul>
  <li>Open RStudio</li>
  <li>Go to File → New File → R Markdown</li>
  <li>Give your document a title, something like “A UN Report on the Relationship between GDP and Life Expectancy” (Note: this is not the same as the file name - it’s just a title that will appear at the top of your report)</li>
  <li>Keep the default output format as HTML.</li>
  <li>R Markdown files always end in <code class="language-plaintext highlighter-rouge">.Rmd</code></li>
</ul>

<blockquote class="callout">
  <h2 id="r-markdown-outputs">R Markdown Outputs</h2>
  <p>The default output for an R Markdown report is HTML, but you can also use R Markdown to <a href="https://rmarkdown.rstudio.com/lesson-9.html">output other report formats</a>. For example, you can generate PDF reports using R Markdown, but you must <a href="https://www.latex-project.org/get/">install TeX</a> to do this.</p>

</blockquote>

<h2 id="basic-components-of-r-markdown">Basic components of R Markdown</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<h3 id="header">Header</h3>
<p><em><a href="#contents">Back to top</a></em></p>

<p>The first part is a <em>header</em> at the top of the file between the lines of <code class="language-plaintext highlighter-rouge">---</code>. This contains instructions for R to specify the type of document to be created and options to choose (ex., title, author, date). These are in the form of key-value pairs (<code class="language-plaintext highlighter-rouge">key: value</code>; YAML).</p>

<p>Here’s an example:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>---
title: 'Writing Reports with R Markdown'
author: "Zena Lapp"
date: "11/18/2020"
output: html_document
---
</code></pre></div></div>

<h3 id="code-chunks">Code chunks</h3>
<p><em><a href="#contents">Back to top</a></em></p>

<p>The next section is a <em>code chunk</em>, or embedded R code, that sets up options for all code chunks. Here is the default when you create a new R Markdown file:</p>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = normalizePath("..")
```
</code></pre></div></div>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
</code></pre></div></div>

<p>All code chunks have this format:</p>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>```{r}
# Your code here
```
</code></pre></div></div>

<p>All of the code is enclosed in 3 back ticks (), and the <code class="language-plaintext highlighter-rouge">{r}</code> part indicates that it’s a chunk of R code.</p>

<p>You can also include other information within the curly brackets to indicate different information about that code chunk.
For instance, the first code block is named “setup”, and <code class="language-plaintext highlighter-rouge">include=FALSE</code> prevents code and results from showing up in the output file.</p>

<p>Inside the code chunk, you can put any R code that you want to run, and you can have as many code chunks as you want in your file.</p>

<p>As we mentioned above, in the first code chunk you set options for the entire file.
<code class="language-plaintext highlighter-rouge">echo = TRUE</code> means that you want your code to be shown in the output file. If you change this to <code class="language-plaintext highlighter-rouge">echo = FALSE</code>, then the code will be hidden and only the output of the code chunks will be seen in the output file.
There are also <a href="https://rmarkdown.rstudio.com/lesson-3.html">many other options that you can change</a>, but we won’t go into those details in this workshop.</p>

<h3 id="text">Text</h3>
<p><em><a href="#contents">Back to top</a></em></p>

<p>Finally, you can include text in your R Markdown file.
This is any text or explanation you want to include, and it’s formatted with Markdown.
We’ll learn more about Markdown formatting soon!</p>

<h2 id="starting-the-report">Starting the report</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>Let’s return to the new R Markdown file you created and delete everything below the setup code chunk. (That stuff is just examples and reminders of how to use R Markdown.)</p>

<p>Next, let’s save our R markdown file to the <code class="language-plaintext highlighter-rouge">reports</code> directory.
You can do this by clicking the save icon in the top left or using <kbd>control</kbd> + <kbd>s</kbd> (<kbd>command</kbd> + <kbd>s</kbd>  on a Mac).</p>

<p>There’s one other thing that we need to do before we get started with our report.
To render our documents into html format, we can “knit” them in R Studio.
Usually, R Markdown renders documents from the directory where the document is saved (the location of the <code class="language-plaintext highlighter-rouge">.Rmd</code> file), but we want it to render from the main project directory where our <code class="language-plaintext highlighter-rouge">.Rproj</code> file is.
This is because that’s where all of our relative paths are from and it’s good practice to have all of your relative paths from the main project directory.
To change this default, click on the down arrow next to the “Knit” button at the top left of R Studio, go to “Knit Directory” and click “Project Directory”.
Now it will assume all of your relative paths for reading and writing files are from the <code class="language-plaintext highlighter-rouge">un-report</code> directory, rather than the <code class="language-plaintext highlighter-rouge">reports</code> directory.</p>

<p>Now that we have that set up, let’s start on the report!</p>

<p>We’re going to use the code you generated yesterday to plot GDP vs. Life Expectancy to include in the report. Recall that we needed a couple R packages to generate these plots. We can create a new code chunk to load the needed packages. You could also include this in the previous setup chunk, it’s up to your personal preference.</p>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>```{r packages}
library(tidyverse)
```
</code></pre></div></div>

<p>Now, in a real report this is when we would type out the background and purpose of our analysis to provide context to our readers. However, since writing is not a focus of this workshop we will avoid lengthy prose and stick to short descriptions. You can copy the following text  into your own report below the package code chunk.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>This report was prepared to the attention of the United Nations. It analyzes the relationship between a country's GDP, its life expectancy and CO2 emissions. Our goal is to determine to what degree a country’s economic strength or weakness may be related to its public health status and impact on climate pollution. We hypothesize that both life expectancy and CO2 emissions increase with a country's GDP.
</code></pre></div></div>

<p>Now, since we want to show our results comparing GDP and life expectancy by country, we need to read in this data so we can regenerate our plot. We will add another code chunk to prepare the data.</p>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>```{r data}
gapminder_1997 &lt;- read_csv("data/gapminder_1997.csv")
```
</code></pre></div></div>

<p>Now that we have the data, we need to produce the plot. Let’s create it!</p>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>```{r gdp_lifeexp_1997}
ggplot(data = gapminder_1997) + 

  aes(x = gdpPercap, y = lifeExp, color=continent, size=pop/1000000) +

  geom_point() +

  labs(x = "GDP Per Capita", y = "Life Expectancy",

       title= "Do people in wealthy countries live longer?", size="Population (in millions)")
```
</code></pre></div></div>

<p>Now we can knit our document to see how our report looks! Use the <kbd>knit</kbd> button in the top left of the screen.</p>

<p><img src="../fig/r-markdown/first_report_render.png" alt="" /></p>

<p>Amazing! We’ve created a report!</p>

<p>It’s looking pretty good, but there seem to be a few extra bits that we don’t need in the report. For example, the report shows that we load the tidyverse package and the accompanying messages.</p>

<p><img src="../fig/r-markdown/tidyverse_messages.png" alt="" /></p>

<p>To get rid of this, we can revise our packages code chunk by adding <code class="language-plaintext highlighter-rouge">include=FALSE</code> just like in the setup chunk to prevent code and messages in this chunk from showing up in our report.</p>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>```{r packages, include=FALSE}
library(tidyverse)
```
</code></pre></div></div>

<p>We can also see the code that was used to generate the plot. Depending on the purpose and audience for your report, you may want to include the code. If you don’t want the code to appear, how can you prevent it? What happens if we add <code class="language-plaintext highlighter-rouge">include=FALSE</code> to the plot code chunk, too? Try rendering the R Markdown report with this change.</p>

<p><img src="../fig/r-markdown/includeFalse.png" alt="" /></p>

<p>Oops! Now the plot doesn’t show up in our report at all. This is because setting <code class="language-plaintext highlighter-rouge">include=FALSE</code> prevents anything in the code chunk from appearing in the report. Instead we can add <code class="language-plaintext highlighter-rouge">echo=FALSE</code> to tell this code chunk that we don’t want to see the code but just the output.</p>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>```{r gdp_lifeexp_1997, echo=FALSE}
ggplot(data = gapminder_1997) + 

  aes(x = gdpPercap, y = lifeExp, color=continent, size=pop/1000000) +

  geom_point() +

  labs(x = "GDP Per Capita", y = "Life Expectancy",

       title= "Do people in wealthy countries live longer?", size="Population (in millions)")
```
</code></pre></div></div>

<p>When we knit this again, our plot is back!</p>

<p><img src="../fig/r-markdown/echoFalse.png" alt="" /></p>

<p>Before we finalize our report, let’s look at a few other cool features. Sometimes, you want to describe your data or results (like our plot) to the audience in text but the data and results may still change as you work things out. R Markdown offers an easy way to do this dynamically, so that the text updates as your data or results change. Here is how to do this.</p>

<p>First, let’s create a code chunk that summarizes features of our data that we can use to describe our plot to  our audience. Note that we set <code class="language-plaintext highlighter-rouge">include=FALSE</code> because we only want this step to happen in the background. For our purposes, we will calculate how many countries were included in the analysis, as well as the minimum and maximum GDP per capita values:</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gapminder_1997</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_csv</span><span class="p">(</span><span class="s2">"data/gapminder_1997.csv"</span><span class="p">)</span><span class="w">

</span><span class="n">nCountries</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">select</span><span class="p">(</span><span class="n">country</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">n_distinct</span><span class="p">()</span><span class="w">

</span><span class="n">minGDP</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarise</span><span class="p">(</span><span class="nf">round</span><span class="p">(</span><span class="nf">min</span><span class="p">(</span><span class="n">gdpPercap</span><span class="p">)))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">pull</span><span class="p">()</span><span class="w">

</span><span class="n">maxGDP</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gapminder_1997</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">summarise</span><span class="p">(</span><span class="nf">round</span><span class="p">(</span><span class="nf">max</span><span class="p">(</span><span class="n">gdpPercap</span><span class="p">)))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">pull</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p>Now, all we need to do is reference the values we just computed to describe our
plot. To do this, we enclose each value in one set of backticks 
(<code class="language-plaintext highlighter-rouge">`r some_R_variable_name `</code>), while the <code class="language-plaintext highlighter-rouge">r</code> part once again
indicates that it’s a chunk of R code. When we knit our report, R will
automatically fill in the values we just created in the above code chunk. Note
that R will automatically update these values every time our data might change
(if we were to decide to drop or add countries to this  analysis, for example).</p>

<div class="language-plaintext code highlighter-rouge"><div class="highlight"><pre class="highlight"><code>The above plot shows the relationship between GDP per capita and life expectancy
for a total of `r nCountries ` countries. For this set of countries,
economic wealth ranged from a minimum of USD `r minGDP`
to a maximum of USD `r maxGDP` per capita.
</code></pre></div></div>

<p>In addition to reporting specific values in the text, we may also want to show a table of values. With R Markdown there are multiple ways to product tables. One way to generate smaller tables is manually. Using a special format we can generate a table in our output. Note that this does not get generated in a code chunk because it is markdown formatting not R code.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>|HEADER 1|HEADER 2|
|-------------|-------------|
|row 1, column1|row 1, column 2|
|row 2, column1|row 2, column 2|
</code></pre></div></div>
<p>Columns are separated by the pipe key <kbd>|</kbd> located above <kbd>Enter</kbd> on the keyboard. The dashes distinguish the header row from the rest of the table. This header could be a name for each column or a header for the whole table. Now that we know the basic structure we can fill in our table. This is how we could present the same numbers from the previous paragraph as a table instead, again using in-line code.
When we knit the report again, the code above will render like this:
<img src="../fig/r-markdown/table_format.png" alt="" /></p>

<p>Here’s the text that we need to include to creata summary table of our data:</p>

<div class="language-plaintext output highlighter-rouge"><div class="highlight"><pre class="highlight"><code>```
|Summary of Data|
|------|------|
|Number of Countries|`r nCountries`|
|Minimum GDP per capita|`r minGDP`|
|Maximum GDP per capita|`r maxGDP`|

```
</code></pre></div></div>

<p>This will render like this:</p>

<p><img src="../fig/r-markdown/table_fillin.png" alt="" /></p>

<p>This is useful if we are reporting a few values, but can get tedious for larger tables. Another way we can add tables to our reports is using an R function called <code class="language-plaintext highlighter-rouge">kable()</code>. Since this is an R function, we will use it within a code chunk. We can give the <code class="language-plaintext highlighter-rouge">kable()</code> function a data table and it will format it to a nice looking table in the report. For example, we could use the following code to generate a table of all the countries in Oceania. The rendered version should look almost exactly as it does on this webpage.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># load library</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">knitr</span><span class="p">)</span><span class="w">

</span><span class="c1"># print kable</span><span class="w">
</span><span class="n">gapminder_1997</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">filter</span><span class="p">(</span><span class="n">continent</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Oceania"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">kable</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<table>
  <thead>
    <tr>
      <th style="text-align: left">country</th>
      <th style="text-align: right">pop</th>
      <th style="text-align: left">continent</th>
      <th style="text-align: right">lifeExp</th>
      <th style="text-align: right">gdpPercap</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">Australia</td>
      <td style="text-align: right">18565243</td>
      <td style="text-align: left">Oceania</td>
      <td style="text-align: right">78.83</td>
      <td style="text-align: right">26997.94</td>
    </tr>
    <tr>
      <td style="text-align: left">New Zealand</td>
      <td style="text-align: right">3676187</td>
      <td style="text-align: left">Oceania</td>
      <td style="text-align: right">77.55</td>
      <td style="text-align: right">21050.41</td>
    </tr>
  </tbody>
</table>

<h2 id="formatting">Formatting</h2>
<p><em><a href="#contents">Back to top</a></em></p>

<p>We now know how to create a report with R Markdown. Maybe we also want to format the report a little bit to structure our thought process in a useful way (e.g., sections) and make it visually appealing?
Markdown is a very simple programming language when it comes to syntax.
Let’s try to figure out some syntax together. Suppose we wanted to create sections in our report.</p>

<blockquote class="challenge">
  <h2 id="r-markdown-headers">R Markdown headers</h2>
  <p>Try googling how to create sections by using headers and subheaders using R Markdown. What do you find?</p>

  <blockquote class="solution">
    <h2 id="solution-2">Solution</h2>
    <p>We can easily create headers and subheaders by using the <code class="language-plaintext highlighter-rouge">#</code> pound/hash sign. Our main headers have one <code class="language-plaintext highlighter-rouge">#</code> (e.g. <code class="language-plaintext highlighter-rouge"># Main Header Here</code>) and to create subheaders we add additinal <code class="language-plaintext highlighter-rouge">#</code>s (e.g. <code class="language-plaintext highlighter-rouge">## First subheader</code> and <code class="language-plaintext highlighter-rouge">### Second subheader</code>)</p>
  </blockquote>
</blockquote>

<p>OK, now that we know how to make headers, let’s practice some more Markdown syntax.</p>

<blockquote class="challenge">
  <h2 id="r-markdown-syntax">R Markdown syntax</h2>
  <p>Go ahead and do some online searches on how to do the following:</p>
  <ul>
    <li>create a bullet point list with three items</li>
    <li>as the first item, write the name of your currently favorite programming language in bold</li>
    <li>as the second item, write  the name of a function you have so far found most useful in italics</li>
    <li>as the third item, write one thing you want to learn next on your programming journey in bold and italics</li>
    <li>turn your bullet point list into a numbered list</li>
    <li>create a fourth list item and find an online guide and/or cheat sheet for basic Markdown syntax, write its name down here and hyperlink its url</li>
  </ul>

  <blockquote class="solution">
    <h2 id="solution-3">Solution</h2>
    <p><a href="https://rmarkdown.rstudio.com/authoring_basics.html">This link</a> has some helpful basic R Markdown syntax.</p>
  </blockquote>
</blockquote>

<h1 id="applying-it-to-your-own-data">Applying it to your own data</h1>

<p>Now it’s time to merge all of your analyses with your own data together into a report. Let us know if you have questions!</p>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>R Markdown is an easy way to create a report that integrates text, code, and figures.</p>
</li>
    
    <li><p>Options such as <code class="language-plaintext highlighter-rouge">include</code> and <code class="language-plaintext highlighter-rouge">echo</code> determine what parts of an R code chunk are included in the R Markdown report.</p>
</li>
    
    <li><p>R Markdown can render HTML, PDF, and Microsoft Word outputs.</p>
</li>
    
  </ul>
</blockquote>

<hr />

<h1 id="conclusion" class="maintitle">Conclusion</h1>

<blockquote class="objectives">
  <h2>Overview</h2>

  <div class="row">
    <div class="col-md-3">
      <strong>Teaching:</strong> 15 min
      <br />
      <strong>Exercises:</strong>  min
    </div>
    <div class="col-md-9">
      <strong>Questions</strong>
      <ul>
	
	<li><p>What do I do after the workshop to apply what I learned and keep learning more?</p>
</li>
	
	<li><p>Where can I learn more coding skills?</p>
</li>
	
	<li><p>How do I deal with coding errors (i.e. debug)?</p>
</li>
	
	<li><p>What other coding concepts should I learn?</p>
</li>
	
      </ul>
    </div>
  </div>

  <div class="row">
    <div class="col-md-3">
    </div>
    <div class="col-md-9">
      <strong>Objectives</strong>
      <ul>
	
	<li><p>Learn how to get help with code via the Internet.</p>
</li>
	
	<li><p>Learn about other coding concepts that would be good to learn in the future.</p>
</li>
	
      </ul>
    </div>
  </div>

</blockquote>

<h1 id="where-to-go-from-here-departing-on-your-own-coding-journey">Where to go from here?: Departing on your own coding journey</h1>

<p><img src="../fig/fig-next-steps.001.jpeg" alt="Learning and debugging throughout the data programming process." /></p>

<p>We have come to the end of this workshop. You learned some basic procedures for importing, managing, visualizing and reporting your data.</p>

<p>As you continue on your coding journey, two things will happen:</p>

<ol>
  <li>You will encounter bugs and need to figure out how to solve them (“debugging”), and</li>
  <li>You will want to learn new data processing and analysis techniques.</li>
</ol>

<p>As we complete the course, we want to  share with you some tips and tricks that have helped us on our own programming journeys.</p>

<h2 id="get-help-and-connect">Get help and connect</h2>

<p>We want to be a resource for you after the workshop ends, and we also want you all to be a resource for each other. Here are ways we can help, let’s discuss what would work best for you to be able to keep in touch and get help/guidance as needed.</p>

<p>First, <strong>you can email us whenever you want with questions!</strong> (zena.lapp@duke.edu; christine.markwalter@duke.edu; erica.zeno@duke.edu)</p>

<p>For additional consistent support, here are some ideas we could implement:</p>

<ol>
  <li>Whatsapp group</li>
  <li>Pro: already commonly used</li>
  <li>Con: can’t create conversation threads and difficult to answer more complicated questions</li>
  <li>Slack</li>
  <li>Pro: can have different channels for different topics, and can use threads for specific conversations</li>
  <li>Con: not as widely used</li>
  <li>Monthly virtual office hours</li>
  <li>Pro: can discuss complex topics and questions more easily</li>
  <li>Con: no immediate help, and might not work with everyone’s schedule</li>
  <li>Follow-up workshops</li>
  <li>Pro: can discuss more detailed topics</li>
  <li>Con: not regular, and different learners may have different specialized topic interests</li>
</ol>

<h2 id="dealing-with-coding-errors">Dealing with coding errors</h2>

<p>Even well seasoned coders run into bugs all the time. Here are some strategies of how programmers try to deal with coding errors:</p>

<ul>
  <li>Don’t panic. Bugs are a normal part of the coding process. Think of each bug as a great opportunity for becoming better at writing code. When you’re debugging, you’re being a detective and trying to figure out what’s wrong. Think of it as a fun mystery to solve! You’ll become a better detective the more you practice.</li>
  <li>First, determine which type of problem you are having: either the code tries to run but returns an error message, or your code runs without error but you did not get the result you expect.</li>
  <li>If you are getting an error message, read the error message carefully. Unfortunately, not all error messages are well written and it may not be obvious at first what is wrong. Try to determine if you are getting a syntax error or a runtime error.
    <ul>
      <li>A <em>syntax error</em> is when the program doesn’t understand how to run the code you provided. In R, these error messages usually contain the word “unexpected”. Often these are because of a typo. Check that your parentheses and quotes are balanced and check that you haven’t misspelled a variable or function name. It’s difficult to identify the exact location where an error starts so you may have to look at lines before the line where the error was reported. In RStudio, look at the code coloring to find anything that looks off. RStudio will also put a red x or an yellow exclamation point to the left of lines where there is a syntax error.</li>
      <li>A <em>runtime error</em> is when the program understood all the commands you wrote, but ran into an error when trying to run them. If you are running multiple commands, try running each on its own. Before each command, check that you are passing the values you expect; and after each command, verify that the results seem sensible.</li>
    </ul>
  </li>
  <li>If you have been able to find parts of your code that cause the error, take some time to review the documentation for the functions or commands you are using. Even for functions or commands you’ve used before, there may still be parts of the documentation that you just haven’t needed to use before. See if there are examples in the documentation that you can compare your own code to. Most programs make finding the documentation pretty easy.
    <ul>
      <li><strong>R</strong>: Typing a question mark immediately followed by a function name, i.e., <code class="language-plaintext highlighter-rouge">?&lt;command&gt;</code> (e.g. ?print), pulls up the R documentation for that function. This provides  a function description, as well as in-depth information about each function argument including its purpose, default and other available options.</li>
      <li><strong>Shell</strong>: Using <code class="language-plaintext highlighter-rouge">man &lt;command&gt;</code> or <code class="language-plaintext highlighter-rouge">&lt;command&gt; --help</code> pulls up the manual page with more information. To quit, click <code class="language-plaintext highlighter-rouge">q</code>.</li>
      <li><strong>Git</strong>: <code class="language-plaintext highlighter-rouge">git help</code> displays help information about git. Read more <a href="https://git-scm.com/docs/git-help">here</a>.</li>
    </ul>
  </li>
  <li>Search online for the error message along with the function that is not working. Most likely, someone encountered this error before you and the examples online will help you figure out your own issue.</li>
  <li>Perform a general online search for what you are trying to do, e.g., “how to import a comma separated spreadsheet into R”, “how to delete a directory from the command line”, “how to merge conflicts in git”. Typically, you will find step-by-step online documentation that you can adapt for your own purposes.</li>
</ul>

<h2 id="debugging-code">Debugging code</h2>

<p>If searching for your particular code problem hasn’t turned up a solution, you may have to do a bit of debugging. <em>Debugging</em> is the process of finding exactly what caused your error, and changing only what is necessary to fix it. There are many strategies to debugging code. Consider checking out the following resources to learn more about it.</p>
<ul>
  <li><a href="https://youtu.be/vLL4mvVL8g0">“5 Essential Tips to Debug Any Piece of Code” by mayuko</a> [video, 8min] - Good general advice for debugging.</li>
  <li><a href="https://rstudio.com/resources/rstudioconf-2020/object-of-type-closure-is-not-subsettable/">“Object of type ‘closure’ is not subsettable” by Jenny Bryan</a> [video, 50min] - A great talk with R specific advice about dealing with errors as a data scientist.</li>
</ul>

<h2 id="asking-strangers-for-help">Asking strangers for help</h2>

<p>If you are unable to determine what’s wrong with your own code, the internet offers several possible ways to get help: asking questions on programming websites, interacting with developers on GitHub, chatting with other programmers on Slack, or reaching out on Twitter. If you’re intimidated by asking people on the internet, you can also reach out to <a href="#get-help-and-connect">people at the University of Michigan</a>. You don’t have to do this all on your own. However, there are some important things to keep in mind when asking questions - whether it be to people on the internet, or to people at the university. You may want to consider these tips to help you increase your chances of getting the support you need:</p>
<ul>
  <li>First, provide a brief summary of what you are trying to accomplish. You may have been working on the problem for a long time but the person you are asking will have no idea of your work until now. Keep it short and focus on your ultimate goal rather than just talking about your code. Sometimes there may better ways to solve the problem than the solution you’ve attempted this far.</li>
  <li>Then, briefly mention what you’ve tried so far. Maybe mention the search terms you used online or the specific documentation you’ve looked at. It should be clear that you’ve at least tried to solve the problem yourself first (after all, that’s the way you’re most likely to learn the most).</li>
  <li>Now, describe the problem you are having. Don’t just say “the code doesn’t work.” Are you getting an error? If so, what is the exact message? Are you getting unexpected results? If so, then what exactly were you expecting to happen?</li>
  <li>Next, include a <a href="https://stackoverflow.com/help/minimal-reproducible-example"><em>minimal, reproducible example</em></a>. This means you need to take away any parts of your code that aren’t necessary to create the error you are dealing with. Often, this may mean starting a branch with a new program just for testing. Make sure the code itself can be run by including or randomly generating some sample data. If there’s no way to copy/paste and run the code on a computer, that means the individuals you are asking for help need to run the code in their head which is a lot harder than running code on a computer. Clearly describe the desired behavior so that possible solutions can be tested and verified. In the process of creating a minimal, reproducible example, it is possible you will solve your own problem.</li>
  <li>Keep in mind the person helping you is unlikely to know the solution off the top of their head. Even the best programmers don’t have all the documentation memorized. But with time, programmers get better at solving problems based on their past experiences. By providing a reproducible example, you allow someone else to use their debugging skills to help you rather than making them guess what might be wrong.</li>
  <li>If it’s not obvious from your question, it can be very useful to include details about the operation system you are using and the versions of the programs you are currently running. Sometimes, things work differently as programs change over time and the code you found on the web may have been for a different version.</li>
  <li>Make it easy to reply. Try to ask one specific question. Asking a bunch of questions takes more time and energy to respond, and you want to minimize the work you create for the kind stranger who is helping you out. Also, if you include sample data with which to test possible solutions, then someone can verify their suggestion works before replying. This will help minimize back-and-forth with your helper which can take time.</li>
  <li>Finally, recognize that asking a good question is an art, not a science. It is something you will get better at with more time. You can also get better by trying to help other people with their questions. That way, you can see what information is necessary to really help someone. It’s a delicate balance between providing enough detail so it’s possible to intelligently respond yet keeping things short enough to not overwhelm your helper (more isn’t always better).</li>
</ul>

<h2 id="learning-new-code">Learning new code</h2>

<p>Free open-source programming languages such as bash, Git and R are constantly evolving. As you try out new data processing and analysis techniques, you will continue to learn new coding logic, concepts, functions, and libraries. Widely available user tools and documentation are a main benefit of free open-source software.</p>

<p>In the following, we list some strategies and resources we find useful. As you move forward, you are likely to identify other resources that fit your own learning style.</p>

<h3 id="general">General</h3>

<ul>
  <li><a href="https://stackoverflow.com/">Stack Overflow</a>: Users post questions or errors and other users respond. (This is a great place to search for help but might not be the best place to ask a question if you are a beginner.)</li>
  <li>Cheat Sheets!! You can find many different cheat sheets for entire languages or specific libraries or even functions online. Just search for “<something> cheat sheet" to see what is available. We have listed a few below.</something></li>
  <li>Study other people’s code. Increasingly, people publish their code (e.g., in a GitHub repository) to facilitate reproducible research. You can take someone else’s code as a starting point and adapt it to your own needs.</li>
  <li>Other people in your department or lab are also great resources for asking about specific coding problems.</li>
</ul>

<h3 id="r">R</h3>

<ul>
  <li><a href="https://community.rstudio.com/">RStudio Community Forum</a>. This is a site that is welcoming to new R users where you can ask R programming questions.</li>
  <li><a href="https://rdrr.io/">Search R Package Documentation</a>. This site is very useful when you see someone using a function but don’t know which package it comes from. This site lets you search across the help pages for all R packages.</li>
  <li><a href="https://rstudio.com/resources/cheatsheets/">RStudio Cheat Sheets</a>. A good collection of cheat sheets to print out and hang at your desk. For example: <a href="https://github.com/rstudio/cheatsheets/blob/main/base-r.pdf">base R</a>, <a href="https://s3.amazonaws.com/assets.datacamp.com/blog_assets/Tidyverse+Cheat+Sheet.pdf">tidyverse</a>, and <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-visualization.pdf">ggplot</a></li>
  <li>If there is a specific R package that you want to use (e.g. <a href="http://www.schlosslab.org/mikropml/">mikropml</a> for machine learning analysis), you can use the vignettes to help you learn the basics of the package. You can search for the vignette online, or use the <code class="language-plaintext highlighter-rouge">vignette()</code> function in R to open them.</li>
</ul>

<h3 id="r-markdown">R Markdown</h3>

<ul>
  <li><a href="https://rmarkdown.rstudio.com/lesson-1.html">Getting started with R Markdown Online Tutorial</a></li>
  <li><a href="https://github.com/rstudio/cheatsheets/blob/main/rmarkdown-2.0.pdf">R Markdown Cheat Sheet</a></li>
  <li><a href="http://www.utstat.toronto.edu/reid/sta2201s/rmarkdown-reference.pdf">R Markdown Reference Guide</a></li>
</ul>

<h2 id="some-important-advanced-coding-concepts-that-you-will-want-to-learn-if-you-continue-coding-a-lot">Some important advanced coding concepts that you will want to learn if you continue coding a lot</h2>

<p>There are some coding concepts that we did not have time to cover in this workshop, but are important to learn as you continue on your journey and begin to perform more sophisticated data analysis projects. While we have not created resources for these topics, we provide some links to where you can learn more. Note that these are more advanced coding topics; you should be come comfortable with what you learned in the workshop before trying to delve deeper into these other concepts. However, you’ll likely come across situations where one of these will be useful, and that’s when you should learn it!</p>

<p>We’ve provided some links below, but feel free to search for other explanations and tutorials as well.</p>

<h3 id="r-coding-topics">R coding topics</h3>

<p>Some more advanced R coding topics include:</p>

<ul>
  <li><a href="https://www.tutorialspoint.com/r/r_lists.htm">Lists</a></li>
  <li><a href="https://swcarpentry.github.io/r-novice-gapminder/10-functions/index.html">Functions</a></li>
  <li><a href="https://swcarpentry.github.io/r-novice-gapminder/07-control-flow/index.html">Conditionals</a></li>
  <li><a href="https://swcarpentry.github.io/r-novice-inflammation/15-supp-loops-in-depth/">Loops and apply statements</a></li>
</ul>

<p><a href="https://erdavenport.github.io/R-ecology-lesson/03-loops-and-functions.html#ifelse_statements">Here</a> is a nice tutorial on conditionals, loops, and functions all together.</p>

<h2 id="domain-specific-analyses">Domain-specific analyses</h2>

<p>We encourage you to investigate domain-specific packages and software that will help you perform specific tasks related to your own research. The best way to find these packages is to either ask other people in your field and/or search for specific tasks that you would like to perform. If you’d like to perform the task in R, include that in your search (e.g. “find pairwise distances for DNA sequences in R” will help you find the R package <code class="language-plaintext highlighter-rouge">ape</code> which has a number of functions to perform phylogenetic and evolutionary analyses in R.)</p>

<blockquote class="keypoints">
  <h2>Key Points</h2>
  <ul>
    
    <li><p>When it comes to trying to figure out how to code something, and debugging, Internet searching is your best friend.</p>
</li>
    
    <li><p>There are several resources at the University of Michigan that you can take advantage of if you need help with your code.</p>
</li>
    
    <li><p>We didn’t have time to cover all important coding concepts in this workshop, so definitely continue trying to learn more once you get comfortable with the material we covered.</p>
</li>
    
    <li><p>There are often packages and tools that you can leverage to perform domain-specific analyses, so search for them!</p>
</li>
    
  </ul>
</blockquote>

<hr />


</article>

      
      






<footer>
  <div class="row">
    <div class="col-md-6 license" id="license-info" align="left">
	
	Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2018–2022
	by <a href="https://carpentries.org/">The Carpentries</a>
        <br>
        Licensed under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a> 2016–2018
	by <a href="https://software-carpentry.org">Software Carpentry Foundation</a>
	
    </div>
    <div class="col-md-6 help-links" align="right">
	
	<a href="/edit//aio.md" data-checker-ignore>Edit on GitHub</a>
	
	/
	<a href="/blob//CONTRIBUTING.md" data-checker-ignore>Contributing</a>
	/
	<a href="/">Source</a>
	/
	<a href="/blob//CITATION" data-checker-ignore>Cite</a>
	/
	<a href="mailto:zena.lapp@duke.edu">Contact</a>
    </div>
  </div>
  <div class="row">
    <div class="col-md-12" align="center">
      Using <a href="https://github.com/carpentries/styles/">The Carpentries style</a>
      version <a href="https://github.com/carpentries/styles/releases/tag/v9.5.3">9.5.3</a>.
    </div>
  </div>
</footer>

      
    </div>
    
<script src="/assets/js/jquery.min.js"></script>
<script src="/assets/js/bootstrap.min.js"></script>
<script src="/assets/js/lesson.js"></script>


<!-- Matomo -->
<script>
  var _paq = window._paq = window._paq || [];
  /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
  _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
  _paq.push(["setDomains", ["*.lessons.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io"]]);
  _paq.push(["setDoNotTrack", true]);
  _paq.push(["disableCookies"]);
  _paq.push(['trackPageView']);
  _paq.push(['enableLinkTracking']);
  (function() {
    var u="https://carpentries.matomo.cloud/";
    _paq.push(['setTrackerUrl', u+'matomo.php']);
    _paq.push(['setSiteId', '1']);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.async=true; g.src='//cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
  })();
</script>
<!-- End Matomo Code -->


<script src="/assets/js/anchor.min.js"></script>
<script>
    anchors.add();
</script>

  </body>
</html>
